// @eslint-react/eslint-plugin/dist/index.js
import { DEFAULT_ESLINT_REACT_SETTINGS } from "@eslint-react/shared";
import reactX from "eslint-plugin-react-x";
import reactDomPlugin from "eslint-plugin-react-dom";
import reactJsxPlugin from "eslint-plugin-react-jsx";
import reactNamingConventionPlugin from "eslint-plugin-react-naming-convention";
import reactRscPlugin from "eslint-plugin-react-rsc";
import reactWebApiPlugin from "eslint-plugin-react-web-api";

//#region \0rolldown/runtime.js
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) {
		__defProp(target, name, {
			get: all[name],
			enumerable: true
		});
	}
	if (!no_symbols) {
		__defProp(target, Symbol.toStringTag, { value: "Module" });
	}
	return target;
};

//#endregion
//#region src/configs/x.ts
var x_exports = /* @__PURE__ */ __exportAll({
	name: () => name$23,
	rules: () => rules$22,
	settings: () => settings$8
});
const name$23 = "@eslint-react/x";
const rules$22 = {
	"@eslint-react/error-boundaries": "error",
	"@eslint-react/exhaustive-deps": "warn",
	"@eslint-react/no-access-state-in-setstate": "error",
	"@eslint-react/no-array-index-key": "warn",
	"@eslint-react/no-children-count": "warn",
	"@eslint-react/no-children-for-each": "warn",
	"@eslint-react/no-children-map": "warn",
	"@eslint-react/no-children-only": "warn",
	"@eslint-react/no-children-to-array": "warn",
	"@eslint-react/no-clone-element": "warn",
	"@eslint-react/no-component-will-mount": "error",
	"@eslint-react/no-component-will-receive-props": "error",
	"@eslint-react/no-component-will-update": "error",
	"@eslint-react/no-context-provider": "warn",
	"@eslint-react/no-create-ref": "error",
	"@eslint-react/no-direct-mutation-state": "error",
	"@eslint-react/no-forward-ref": "warn",
	"@eslint-react/no-missing-key": "error",
	"@eslint-react/no-nested-component-definitions": "error",
	"@eslint-react/no-nested-lazy-component-declarations": "error",
	"@eslint-react/no-set-state-in-component-did-mount": "warn",
	"@eslint-react/no-set-state-in-component-did-update": "warn",
	"@eslint-react/no-set-state-in-component-will-update": "warn",
	"@eslint-react/no-unnecessary-use-prefix": "warn",
	"@eslint-react/no-unsafe-component-will-mount": "warn",
	"@eslint-react/no-unsafe-component-will-receive-props": "warn",
	"@eslint-react/no-unsafe-component-will-update": "warn",
	"@eslint-react/no-unused-class-component-members": "warn",
	"@eslint-react/no-use-context": "warn",
	"@eslint-react/purity": "warn",
	"@eslint-react/rules-of-hooks": "error",
	"@eslint-react/set-state-in-effect": "warn",
	"@eslint-react/set-state-in-render": "error",
	"@eslint-react/static-components": "error",
	"@eslint-react/unsupported-syntax": "error",
	"@eslint-react/use-memo": "error",
	"@eslint-react/use-state": "warn"
};
const settings$8 = { "react-x": DEFAULT_ESLINT_REACT_SETTINGS };

//#endregion
//#region src/configs/all.ts
var all_exports = /* @__PURE__ */ __exportAll({
	name: () => name$22,
	rules: () => rules$21,
	settings: () => settings$7
});
const name$22 = "@eslint-react/all";
const rules$21 = {
	"@eslint-react/error-boundaries": "error",
	"@eslint-react/exhaustive-deps": "warn",
	"@eslint-react/globals": "warn",
	"@eslint-react/immutability": "error",
	"@eslint-react/no-access-state-in-setstate": "error",
	"@eslint-react/no-array-index-key": "warn",
	"@eslint-react/no-children-count": "warn",
	"@eslint-react/no-children-for-each": "warn",
	"@eslint-react/no-children-map": "warn",
	"@eslint-react/no-children-only": "warn",
	"@eslint-react/no-children-to-array": "warn",
	"@eslint-react/no-class-component": "warn",
	"@eslint-react/no-clone-element": "warn",
	"@eslint-react/no-component-will-mount": "error",
	"@eslint-react/no-component-will-receive-props": "error",
	"@eslint-react/no-component-will-update": "error",
	"@eslint-react/no-context-provider": "warn",
	"@eslint-react/no-create-ref": "error",
	"@eslint-react/no-direct-mutation-state": "error",
	"@eslint-react/no-duplicate-key": "error",
	"@eslint-react/no-forward-ref": "warn",
	"@eslint-react/no-missing-component-display-name": "warn",
	"@eslint-react/no-missing-context-display-name": "warn",
	"@eslint-react/no-missing-key": "error",
	"@eslint-react/no-misused-capture-owner-stack": "error",
	"@eslint-react/no-nested-component-definitions": "error",
	"@eslint-react/no-nested-lazy-component-declarations": "error",
	"@eslint-react/no-set-state-in-component-did-mount": "warn",
	"@eslint-react/no-set-state-in-component-did-update": "warn",
	"@eslint-react/no-set-state-in-component-will-update": "warn",
	"@eslint-react/no-unnecessary-use-prefix": "warn",
	"@eslint-react/no-unsafe-component-will-mount": "warn",
	"@eslint-react/no-unsafe-component-will-receive-props": "warn",
	"@eslint-react/no-unsafe-component-will-update": "warn",
	"@eslint-react/no-unstable-context-value": "warn",
	"@eslint-react/no-unstable-default-props": "warn",
	"@eslint-react/no-unused-class-component-members": "warn",
	"@eslint-react/no-unused-state": "warn",
	"@eslint-react/no-use-context": "warn",
	"@eslint-react/purity": "warn",
	"@eslint-react/refs": "error",
	"@eslint-react/rules-of-hooks": "error",
	"@eslint-react/set-state-in-effect": "warn",
	"@eslint-react/set-state-in-render": "error",
	"@eslint-react/static-components": "error",
	"@eslint-react/unsupported-syntax": "error",
	"@eslint-react/use-memo": "error",
	"@eslint-react/use-state": "warn",
	"@eslint-react/jsx-no-children-prop": "warn",
	"@eslint-react/jsx-no-children-prop-with-children": "error",
	"@eslint-react/jsx-no-comment-textnodes": "warn",
	"@eslint-react/jsx-no-key-after-spread": "error",
	"@eslint-react/jsx-no-leaked-dollar": "warn",
	"@eslint-react/jsx-no-leaked-semicolon": "warn",
	"@eslint-react/jsx-no-namespace": "error",
	"@eslint-react/jsx-no-useless-fragment": "warn",
	"@eslint-react/rsc-function-definition": "error",
	"@eslint-react/dom-no-dangerously-set-innerhtml": "warn",
	"@eslint-react/dom-no-dangerously-set-innerhtml-with-children": "error",
	"@eslint-react/dom-no-find-dom-node": "error",
	"@eslint-react/dom-no-flush-sync": "error",
	"@eslint-react/dom-no-hydrate": "error",
	"@eslint-react/dom-no-missing-button-type": "warn",
	"@eslint-react/dom-no-missing-iframe-sandbox": "warn",
	"@eslint-react/dom-no-render": "error",
	"@eslint-react/dom-no-render-return-value": "error",
	"@eslint-react/dom-no-script-url": "warn",
	"@eslint-react/dom-no-string-style-prop": "error",
	"@eslint-react/dom-no-unknown-property": "warn",
	"@eslint-react/dom-no-unsafe-iframe-sandbox": "warn",
	"@eslint-react/dom-no-unsafe-target-blank": "warn",
	"@eslint-react/dom-no-use-form-state": "warn",
	"@eslint-react/dom-no-void-elements-with-children": "error",
	"@eslint-react/web-api-no-leaked-event-listener": "warn",
	"@eslint-react/web-api-no-leaked-fetch": "warn",
	"@eslint-react/web-api-no-leaked-interval": "warn",
	"@eslint-react/web-api-no-leaked-resize-observer": "warn",
	"@eslint-react/web-api-no-leaked-timeout": "warn",
	"@eslint-react/naming-convention-context-name": "warn",
	"@eslint-react/naming-convention-id-name": "warn",
	"@eslint-react/naming-convention-ref-name": "warn"
};
const settings$7 = { ...settings$8 };

//#endregion
//#region src/configs/disable-conflict-eslint-plugin-react.ts
var disable_conflict_eslint_plugin_react_exports = /* @__PURE__ */ __exportAll({
	name: () => name$21,
	rules: () => rules$20
});
const name$21 = "@eslint-react/disable-conflict-eslint-plugin-react";
const rules$20 = reactX.configs["disable-conflict-eslint-plugin-react"].rules ?? {};

//#endregion
//#region src/configs/disable-conflict-eslint-plugin-react-hooks.ts
var disable_conflict_eslint_plugin_react_hooks_exports = /* @__PURE__ */ __exportAll({
	name: () => name$20,
	rules: () => rules$19
});
const name$20 = "@eslint-react/disable-conflict-eslint-plugin-react-hooks";
const rules$19 = reactX.configs["disable-conflict-eslint-plugin-react-hooks"].rules ?? {};

//#endregion
//#region src/configs/dom.ts
var dom_exports = /* @__PURE__ */ __exportAll({
	name: () => name$19,
	rules: () => rules$18
});
const name$19 = "@eslint-react/dom";
const rules$18 = {
	"@eslint-react/dom-no-dangerously-set-innerhtml": "warn",
	"@eslint-react/dom-no-dangerously-set-innerhtml-with-children": "error",
	"@eslint-react/dom-no-find-dom-node": "error",
	"@eslint-react/dom-no-flush-sync": "error",
	"@eslint-react/dom-no-hydrate": "error",
	"@eslint-react/dom-no-render": "error",
	"@eslint-react/dom-no-render-return-value": "error",
	"@eslint-react/dom-no-script-url": "warn",
	"@eslint-react/dom-no-unsafe-iframe-sandbox": "warn",
	"@eslint-react/dom-no-use-form-state": "error",
	"@eslint-react/dom-no-void-elements-with-children": "error"
};

//#endregion
//#region src/configs/disable-dom.ts
var disable_dom_exports = /* @__PURE__ */ __exportAll({
	name: () => name$18,
	rules: () => rules$17
});
const name$18 = "@eslint-react/disable-dom";
const rules$17 = Object.fromEntries(Object.entries(rules$18).map(([key]) => [key, "off"]));

//#endregion
//#region src/configs/disable-experimental.ts
var disable_experimental_exports = /* @__PURE__ */ __exportAll({
	name: () => name$17,
	rules: () => rules$16
});
const name$17 = "@eslint-react/disable-experimental";
const rules$16 = {
	"@eslint-react/globals": "off",
	"@eslint-react/immutability": "off",
	"@eslint-react/no-duplicate-key": "off",
	"@eslint-react/no-implicit-children": "off",
	"@eslint-react/no-implicit-key": "off",
	"@eslint-react/no-implicit-ref": "off",
	"@eslint-react/no-misused-capture-owner-stack": "off",
	"@eslint-react/no-unused-props": "off",
	"@eslint-react/no-unused-state": "off",
	"@eslint-react/refs": "off",
	"@eslint-react/set-state-in-render": "off",
	"@eslint-react/static-components": "off",
	"@eslint-react/rsc-function-definition": "off",
	"@eslint-react/web-api-no-leaked-fetch": "off"
};

//#endregion
//#region src/configs/jsx.ts
var jsx_exports = /* @__PURE__ */ __exportAll({
	name: () => name$16,
	rules: () => rules$15
});
const name$16 = "@eslint-react/jsx";
const rules$15 = {
	"@eslint-react/jsx-no-children-prop": "warn",
	"@eslint-react/jsx-no-children-prop-with-children": "error",
	"@eslint-react/jsx-no-comment-textnodes": "warn",
	"@eslint-react/jsx-no-key-after-spread": "error",
	"@eslint-react/jsx-no-leaked-dollar": "warn",
	"@eslint-react/jsx-no-leaked-semicolon": "warn",
	"@eslint-react/jsx-no-namespace": "error"
};

//#endregion
//#region src/configs/disable-jsx.ts
var disable_jsx_exports = /* @__PURE__ */ __exportAll({
	name: () => name$15,
	rules: () => rules$14
});
const name$15 = "@eslint-react/disable-jsx";
const rules$14 = Object.fromEntries(Object.entries(rules$15).map(([key]) => [key, "off"]));

//#endregion
//#region src/configs/naming-convention.ts
var naming_convention_exports = /* @__PURE__ */ __exportAll({
	name: () => name$14,
	rules: () => rules$13
});
const name$14 = "@eslint-react/naming-convention";
const rules$13 = {
	"@eslint-react/naming-convention-context-name": "warn",
	"@eslint-react/naming-convention-id-name": "warn",
	"@eslint-react/naming-convention-ref-name": "warn"
};

//#endregion
//#region src/configs/disable-naming-convention.ts
var disable_naming_convention_exports = /* @__PURE__ */ __exportAll({
	name: () => name$13,
	rules: () => rules$12
});
const name$13 = "@eslint-react/disable-naming-convention";
const rules$12 = Object.fromEntries(Object.entries(rules$13).map(([key]) => [key, "off"]));

//#endregion
//#region src/configs/rsc.ts
var rsc_exports = /* @__PURE__ */ __exportAll({
	name: () => name$12,
	rules: () => rules$11
});
const name$12 = "@eslint-react/rsc";
const rules$11 = { "@eslint-react/rsc-function-definition": "error" };

//#endregion
//#region src/configs/disable-rsc.ts
var disable_rsc_exports = /* @__PURE__ */ __exportAll({
	name: () => name$11,
	rules: () => rules$10
});
const name$11 = "@eslint-react/disable-rsc";
const rules$10 = Object.fromEntries(Object.entries(rules$11).map(([key]) => [key, "off"]));

//#endregion
//#region src/configs/disable-type-checked.ts
var disable_type_checked_exports = /* @__PURE__ */ __exportAll({
	name: () => name$10,
	rules: () => rules$9
});
const name$10 = "@eslint-react/disable-type-checked";
const rules$9 = {
	"@eslint-react/no-implicit-children": "off",
	"@eslint-react/no-implicit-key": "off",
	"@eslint-react/no-implicit-ref": "off",
	"@eslint-react/no-leaked-conditional-rendering": "off",
	"@eslint-react/no-unused-props": "off"
};

//#endregion
//#region src/configs/web-api.ts
var web_api_exports = /* @__PURE__ */ __exportAll({
	name: () => name$9,
	rules: () => rules$8,
	settings: () => settings$6
});
const name$9 = "@eslint-react/web-api";
const rules$8 = {
	"@eslint-react/web-api-no-leaked-event-listener": "warn",
	"@eslint-react/web-api-no-leaked-fetch": "warn",
	"@eslint-react/web-api-no-leaked-interval": "warn",
	"@eslint-react/web-api-no-leaked-resize-observer": "warn",
	"@eslint-react/web-api-no-leaked-timeout": "warn"
};
const settings$6 = { ...settings$8 };

//#endregion
//#region src/configs/disable-web-api.ts
var disable_web_api_exports = /* @__PURE__ */ __exportAll({
	name: () => name$8,
	rules: () => rules$7
});
const name$8 = "@eslint-react/disable-web-api";
const rules$7 = Object.fromEntries(Object.entries(rules$8).map(([key]) => [key, "off"]));

//#endregion
//#region src/configs/off.ts
var off_exports = /* @__PURE__ */ __exportAll({
	name: () => name$7,
	rules: () => rules$6
});
const name$7 = "@eslint-react/off";
const rules$6 = {
	...Object.fromEntries(Object.entries(rules$21).map(([key]) => [key, "off"])),
	...rules$9
};

//#endregion
//#region src/configs/recommended.ts
var recommended_exports = /* @__PURE__ */ __exportAll({
	name: () => name$6,
	rules: () => rules$5,
	settings: () => settings$5
});
const name$6 = "@eslint-react/recommended";
const rules$5 = {
	...rules$22,
	...rules$15,
	...rules$11,
	...rules$18,
	...rules$8,
	...rules$13
};
const settings$5 = { ...settings$8 };

//#endregion
//#region src/configs/recommended-typescript.ts
var recommended_typescript_exports = /* @__PURE__ */ __exportAll({
	name: () => name$5,
	rules: () => rules$4,
	settings: () => settings$4
});
const name$5 = "@eslint-react/recommended-typescript";
const rules$4 = { ...rules$5 };
const settings$4 = { ...settings$5 };

//#endregion
//#region src/configs/recommended-type-checked.ts
var recommended_type_checked_exports = /* @__PURE__ */ __exportAll({
	name: () => name$4,
	rules: () => rules$3,
	settings: () => settings$3
});
const name$4 = "@eslint-react/recommended-type-checked";
const rules$3 = {
	...rules$4,
	"@eslint-react/no-leaked-conditional-rendering": "error"
};
const settings$3 = { ...settings$4 };

//#endregion
//#region src/configs/strict.ts
var strict_exports = /* @__PURE__ */ __exportAll({
	name: () => name$3,
	rules: () => rules$2,
	settings: () => settings$2
});
const name$3 = "@eslint-react/strict";
const rules$2 = {
	...rules$5,
	"@eslint-react/no-class-component": "error",
	"@eslint-react/no-misused-capture-owner-stack": "error",
	"@eslint-react/no-unstable-context-value": "warn",
	"@eslint-react/no-unstable-default-props": "warn",
	"@eslint-react/jsx-no-children-prop": "error",
	"@eslint-react/jsx-no-useless-fragment": "warn",
	"@eslint-react/dom-no-missing-button-type": "warn",
	"@eslint-react/dom-no-missing-iframe-sandbox": "warn",
	"@eslint-react/dom-no-unsafe-target-blank": "warn"
};
const settings$2 = { ...settings$5 };

//#endregion
//#region src/configs/strict-typescript.ts
var strict_typescript_exports = /* @__PURE__ */ __exportAll({
	name: () => name$2,
	rules: () => rules$1,
	settings: () => settings$1
});
const name$2 = "@eslint-react/strict-typescript";
const rules$1 = { ...rules$2 };
const settings$1 = { ...settings$2 };

//#endregion
//#region src/configs/strict-type-checked.ts
var strict_type_checked_exports = /* @__PURE__ */ __exportAll({
	name: () => name$1,
	rules: () => rules,
	settings: () => settings
});
const name$1 = "@eslint-react/strict-type-checked";
const rules = {
	...rules$1,
	"@eslint-react/no-leaked-conditional-rendering": "error",
	"@eslint-react/no-unused-props": "warn"
};
const settings = { ...settings$1 };

//#endregion
//#region package.json
var name = "@eslint-react/eslint-plugin";
var version = "5.8.10";

//#endregion
//#region src/utils/type-of.ts
/**
* This is an enhanced version of the typeof operator to check the type of more complex values.
* In this case we just mind about arrays and objects. We can add more on demand.
* @param t the value to be checked
* @returns the type of the value
*/
function typeOf(t) {
	return Object.prototype.toString.call(t).replace(/^\[object (.+)\]$/, "$1").toLowerCase();
}

//#endregion
//#region src/utils/transform-keys.ts
/**
* This function is used to shallowly transform the keys of an object.
* It will only be transformed at runtime, so it's not type safe.
* @param obj the object to transform.
* @param transform the function to transform the keys from string to string.
* @returns the transformed object.
* @example transformKeys({ 'foo-bar': { 'fizz-buzz': true } }, camelCase)
* // { fooBar: { 'fizz-buzz': true } }
*/
function transformKeys(obj, transform) {
	if (typeOf(obj) !== "object") return obj;
	const res = {};
	for (const key in obj) res[transform(key)] = obj[key];
	return res;
}

//#endregion
//#region src/utils/pad-keys-left.ts
function padKeysLeft(obj, left) {
	return transformKeys(obj, (key) => `${left}${key}`);
}

//#endregion
//#region src/plugin.ts
const plugin = {
	meta: {
		name,
		version
	},
	rules: {
		...reactX.rules,
		...padKeysLeft(reactX.rules, "x-"),
		...padKeysLeft(reactJsxPlugin.rules, "jsx-"),
		...padKeysLeft(reactRscPlugin.rules, "rsc-"),
		...padKeysLeft(reactDomPlugin.rules, "dom-"),
		...padKeysLeft(reactWebApiPlugin.rules, "web-api-"),
		...padKeysLeft(reactNamingConventionPlugin.rules, "naming-convention-")
	}
};

//#endregion
//#region src/index.ts
function createConfig(base) {
	return {
		...base,
		plugins: {
			...base.plugins,
			"@eslint-react": plugin
		}
	};
}
const finalPlugin = {
	...plugin,
	configs: {
		["all"]: createConfig(all_exports),
		["disable-conflict-eslint-plugin-react"]: createConfig(disable_conflict_eslint_plugin_react_exports),
		["disable-conflict-eslint-plugin-react-hooks"]: createConfig(disable_conflict_eslint_plugin_react_hooks_exports),
		["disable-dom"]: createConfig(disable_dom_exports),
		["disable-experimental"]: createConfig(disable_experimental_exports),
		["disable-jsx"]: createConfig(disable_jsx_exports),
		["disable-naming-convention"]: createConfig(disable_naming_convention_exports),
		["disable-rsc"]: createConfig(disable_rsc_exports),
		["disable-type-checked"]: createConfig(disable_type_checked_exports),
		["disable-web-api"]: createConfig(disable_web_api_exports),
		["dom"]: createConfig(dom_exports),
		["jsx"]: createConfig(jsx_exports),
		["naming-convention"]: createConfig(naming_convention_exports),
		["off"]: createConfig(off_exports),
		["recommended"]: createConfig(recommended_exports),
		["recommended-type-checked"]: createConfig(recommended_type_checked_exports),
		["recommended-typescript"]: createConfig(recommended_typescript_exports),
		["rsc"]: createConfig(rsc_exports),
		["strict"]: createConfig(strict_exports),
		["strict-type-checked"]: createConfig(strict_type_checked_exports),
		["strict-typescript"]: createConfig(strict_typescript_exports),
		["web-api"]: createConfig(web_api_exports),
		["x"]: createConfig(x_exports)
	}
};

//#endregion
export { finalPlugin as default };
