// @eslint-react/core/dist/index.js
import { Check, Extract, Traverse, is, isOneOf } from "@eslint-react/ast";
import { resolve, resolveImportSource } from "@eslint-react/var";
import { AST_NODE_TYPES } from "@typescript-eslint/types";
import { randomBytes } from "node:crypto";
import { JsxDetectionHint as JsxDetectionHint$1 } from "@eslint-react/jsx";
import { RE_ANNOTATION_JSX, RE_ANNOTATION_JSX_FRAG, RE_ANNOTATION_JSX_IMPORT_SOURCE, RE_ANNOTATION_JSX_RUNTIME, RE_COMPONENT_NAME, RE_COMPONENT_NAME_LOOSE } from "@eslint-react/shared";
import ts from "typescript";
import { P, isMatching, match } from "ts-pattern";

//#region ../../.pkgs/eff/dist/index.js
/**
* Creates a function that can be used in a data-last (aka `pipe`able) or
* data-first style.
*
* The first parameter to `dual` is either the arity of the uncurried function
* or a predicate that determines if the function is being used in a data-first
* or data-last style.
*
* Using the arity is the most common use case, but there are some cases where
* you may want to use a predicate. For example, if you have a function that
* takes an optional argument, you can use a predicate to determine if the
* function is being used in a data-first or data-last style.
*
* You can pass either the arity of the uncurried function or a predicate
* which determines if the function is being used in a data-first or
* data-last style.
*
* **Example** (Using arity to determine data-first or data-last style)
*
* ```ts
* import { dual, pipe } from "effect/Function"
*
* const sum = dual<
*   (that: number) => (self: number) => number,
*   (self: number, that: number) => number
* >(2, (self, that) => self + that)
*
* console.log(sum(2, 3)) // 5
* console.log(pipe(2, sum(3))) // 5
* ```
*
* **Example** (Using call signatures to define the overloads)
*
* ```ts
* import { dual, pipe } from "effect/Function"
*
* const sum: {
*   (that: number): (self: number) => number
*   (self: number, that: number): number
* } = dual(2, (self: number, that: number): number => self + that)
*
* console.log(sum(2, 3)) // 5
* console.log(pipe(2, sum(3))) // 5
* ```
*
* **Example** (Using a predicate to determine data-first or data-last style)
*
* ```ts
* import { dual, pipe } from "effect/Function"
*
* const sum = dual<
*   (that: number) => (self: number) => number,
*   (self: number, that: number) => number
* >(
*   (args) => args.length === 2,
*   (self, that) => self + that
* )
*
* console.log(sum(2, 3)) // 5
* console.log(pipe(2, sum(3))) // 5
* ```
*
* @param arity - The arity of the uncurried function or a predicate that determines if the function is being used in a data-first or data-last style.
* @param body - The function to be curried.
* @since 1.0.0
*/
const dual = function(arity, body) {
	if (typeof arity === "function") return function() {
		return arity(arguments) ? body.apply(this, arguments) : ((self) => body(self, ...arguments));
	};
	switch (arity) {
		case 0:
		case 1: throw new RangeError(`Invalid arity ${arity}`);
		case 2: return function(a, b) {
			if (arguments.length >= 2) return body(a, b);
			return function(self) {
				return body(self, a);
			};
		};
		case 3: return function(a, b, c) {
			if (arguments.length >= 3) return body(a, b, c);
			return function(self) {
				return body(self, a, b);
			};
		};
		default: return function() {
			if (arguments.length >= arity) return body.apply(this, arguments);
			const args = arguments;
			return function(self) {
				return body(self, ...args);
			};
		};
	}
};
/**
* Do nothing and return `false`.
*
* @returns false
*/
function constFalse() {
	return false;
}
/**
* Composes two functions, `ab` and `bc` into a single function that takes in an argument `a` of type `A` and returns a result of type `C`.
* The result is obtained by first applying the `ab` function to `a` and then applying the `bc` function to the result of `ab`.
*
* @param self - The first function to apply (or the composed function in data-last style).
* @param bc - The second function to apply.
* @returns A composed function that applies both functions in sequence.
* @example
* ```ts
* import * as assert from "node:assert"
* import { compose } from "effect/Function"
*
* const increment = (n: number) => n + 1;
* const square = (n: number) => n * n;
*
* assert.strictEqual(compose(increment, square)(2), 9);
* ```
*
* @since 1.0.0
*/
const compose = dual(2, (ab, bc) => (a) => bc(ab(a)));

//#endregion
//#region src/api.ts
/**
* Check if the node is a React API identifier or member expression
* @param api The React API name to check against (ex: "useState", "React.memo")
* @returns A predicate function to check if a node matches the API
*/
function isAPI(api) {
	const func = (context, node) => {
		if (node == null) return false;
		const expr = Extract.unwrap(node);
		const getText = (n) => context.sourceCode.getText(n);
		const name = Extract.getFullyQualifiedName(expr, getText);
		if (name === api) return true;
		if (name.endsWith(`.${api}`)) return true;
		return false;
	};
	return dual(2, func);
}
/**
* Check if the node is a call expression to a specific React API
* @param api The React API name to check against
* @returns A predicate function to check if a node is a call to the API
*/
function isAPICall(api) {
	const func = (context, node) => {
		if (node == null) return false;
		if (node.type !== AST_NODE_TYPES.CallExpression) return false;
		return isAPI(api)(context, Extract.unwrap(node.callee));
	};
	return dual(2, func);
}
/**
* Check if a variable is initialized from React import
* @param name The variable name
* @param initialScope The initial scope
* @param importSource Alternative import source of React (ex: "preact/compat")
* @returns True if the variable is initialized or derived from React import
*/
function isAPIFromReact(name, initialScope, importSource = "react") {
	return name.toLowerCase() === "react" || Boolean(resolveImportSource(name, initialScope)?.startsWith(importSource));
}
/**
* if a variable is initialized from React Native import
* @param name The variable name
* @param initialScope The initial scope
* @param importSource Alternative import source of React Native (ex: "react-native-web")
* @returns True if the variable is initialized from React Native import
*/
function isAPIFromReactNative(name, initialScope, importSource = "react-native") {
	return [
		"react_native",
		"reactnative",
		"rn"
	].includes(name.toLowerCase()) || Boolean(resolveImportSource(name, initialScope)?.startsWith(importSource));
}
const isCaptureOwnerStack = isAPI("captureOwnerStack");
const isChildrenCount = isAPI("Children.count");
const isChildrenForEach = isAPI("Children.forEach");
const isChildrenMap = isAPI("Children.map");
const isChildrenOnly = isAPI("Children.only");
const isChildrenToArray = isAPI("Children.toArray");
const isCloneElement = isAPI("cloneElement");
const isCreateContext = isAPI("createContext");
const isCreateElement = isAPI("createElement");
const isCreateRef = isAPI("createRef");
const isForwardRef = isAPI("forwardRef");
const isMemo = isAPI("memo");
const isLazy = isAPI("lazy");
const isCaptureOwnerStackCall = isAPICall("captureOwnerStack");
const isChildrenCountCall = isAPICall("Children.count");
const isChildrenForEachCall = isAPICall("Children.forEach");
const isChildrenMapCall = isAPICall("Children.map");
const isChildrenOnlyCall = isAPICall("Children.only");
const isChildrenToArrayCall = isAPICall("Children.toArray");
const isCloneElementCall = isAPICall("cloneElement");
const isCreateContextCall = isAPICall("createContext");
const isCreateElementCall = isAPICall("createElement");
const isCreateRefCall = isAPICall("createRef");
const isForwardRefCall = isAPICall("forwardRef");
const isMemoCall = isAPICall("memo");
const isLazyCall = isAPICall("lazy");
const isUse = isAPI("use");
const isUseActionState = isAPI("useActionState");
const isUseCallback = isAPI("useCallback");
const isUseContext = isAPI("useContext");
const isUseDebugValue = isAPI("useDebugValue");
const isUseDeferredValue = isAPI("useDeferredValue");
const isUseEffect = isAPI("useEffect");
const isUseFormStatus = isAPI("useFormStatus");
const isUseId = isAPI("useId");
const isUseImperativeHandle = isAPI("useImperativeHandle");
const isUseInsertionEffect = isAPI("useInsertionEffect");
const isUseLayoutEffect = isAPI("useLayoutEffect");
const isUseMemo = isAPI("useMemo");
const isUseOptimistic = isAPI("useOptimistic");
const isUseReducer = isAPI("useReducer");
const isUseRef = isAPI("useRef");
const isUseState = isAPI("useState");
const isUseSyncExternalStore = isAPI("useSyncExternalStore");
const isUseTransition = isAPI("useTransition");
const isUseCall = isAPICall("use");
const isUseActionStateCall = isAPICall("useActionState");
const isUseCallbackCall = isAPICall("useCallback");
const isUseContextCall = isAPICall("useContext");
const isUseDebugValueCall = isAPICall("useDebugValue");
const isUseDeferredValueCall = isAPICall("useDeferredValue");
const isUseEffectCall = isAPICall("useEffect");
const isUseFormStatusCall = isAPICall("useFormStatus");
const isUseIdCall = isAPICall("useId");
const isUseImperativeHandleCall = isAPICall("useImperativeHandle");
const isUseInsertionEffectCall = isAPICall("useInsertionEffect");
const isUseLayoutEffectCall = isAPICall("useLayoutEffect");
const isUseMemoCall = isAPICall("useMemo");
const isUseOptimisticCall = isAPICall("useOptimistic");
const isUseReducerCall = isAPICall("useReducer");
const isUseRefCall = isAPICall("useRef");
const isUseStateCall = isAPICall("useState");
const isUseSyncExternalStoreCall = isAPICall("useSyncExternalStore");
const isUseTransitionCall = isAPICall("useTransition");

//#endregion
//#region src/class.ts
/**
* Get the class identifier of a class node
* @param node The class node to get the identifier from
* @returns The class identifier or null if not found
*/
function getClassId(node) {
	if (node.id != null) return node.id;
	if (node.parent.type === AST_NODE_TYPES.VariableDeclarator) return node.parent.id;
	return null;
}

//#endregion
//#region src/class-component.ts
function isClassComponent(node) {
	if ("superClass" in node && node.superClass != null) {
		const re = /^(?:Pure)?Component$/u;
		switch (true) {
			case node.superClass.type === AST_NODE_TYPES.Identifier: return re.test(node.superClass.name);
			case node.superClass.type === AST_NODE_TYPES.MemberExpression && node.superClass.property.type === AST_NODE_TYPES.Identifier: return re.test(node.superClass.property.name);
		}
	}
	return false;
}
/**
* @param node The AST node to check.
* @deprecated Class components are legacy. This function exists only to support legacy rules.
*/
function isPureComponent(node) {
	if ("superClass" in node && node.superClass != null) {
		const re = /^PureComponent$/u;
		switch (true) {
			case node.superClass.type === AST_NODE_TYPES.Identifier: return re.test(node.superClass.name);
			case node.superClass.type === AST_NODE_TYPES.MemberExpression && node.superClass.property.type === AST_NODE_TYPES.Identifier: return re.test(node.superClass.property.name);
		}
	}
	return false;
}
function createLifecycleChecker(methodName, isStatic = false) {
	return (node) => Check.isPropertyOrMethod(node) && node.static === isStatic && node.key.type === AST_NODE_TYPES.Identifier && node.key.name === methodName;
}
/** @deprecated Class components are legacy. */
const isRender = createLifecycleChecker("render");
/** @deprecated Class components are legacy. */
const isComponentDidCatch = createLifecycleChecker("componentDidCatch");
/** @deprecated Class components are legacy. */
const isComponentDidMount = createLifecycleChecker("componentDidMount");
/** @deprecated Class components are legacy. */
const isComponentDidUpdate = createLifecycleChecker("componentDidUpdate");
/** @deprecated Class components are legacy. */
const isComponentWillMount = createLifecycleChecker("componentWillMount");
/** @deprecated Class components are legacy. */
const isComponentWillReceiveProps = createLifecycleChecker("componentWillReceiveProps");
/** @deprecated Class components are legacy. */
const isComponentWillUnmount = createLifecycleChecker("componentWillUnmount");
/** @deprecated Class components are legacy. */
const isComponentWillUpdate = createLifecycleChecker("componentWillUpdate");
/** @deprecated Class components are legacy. */
const isGetChildContext = createLifecycleChecker("getChildContext");
/** @deprecated Class components are legacy. */
const isGetInitialState = createLifecycleChecker("getInitialState");
/** @deprecated Class components are legacy. */
const isGetSnapshotBeforeUpdate = createLifecycleChecker("getSnapshotBeforeUpdate");
/** @deprecated Class components are legacy. */
const isShouldComponentUpdate = createLifecycleChecker("shouldComponentUpdate");
/** @deprecated Class components are legacy. */
const isUnsafeComponentWillMount = createLifecycleChecker("UNSAFE_componentWillMount");
/** @deprecated Class components are legacy. */
const isUnsafeComponentWillReceiveProps = createLifecycleChecker("UNSAFE_componentWillReceiveProps");
/** @deprecated Class components are legacy. */
const isUnsafeComponentWillUpdate = createLifecycleChecker("UNSAFE_componentWillUpdate");
/** @deprecated Class components are legacy. */
const isGetDefaultProps = createLifecycleChecker("getDefaultProps", true);
/** @deprecated Class components are legacy. */
const isGetDerivedStateFromProps = createLifecycleChecker("getDerivedStateFromProps", true);
/** @deprecated Class components are legacy. */
const isGetDerivedStateFromError = createLifecycleChecker("getDerivedStateFromError", true);
/**
* @param node The AST node to check.
* @deprecated Class components are legacy. This function exists only to support legacy rules.
*/
function isRenderMethodLike(node) {
	return Check.isPropertyOrMethod(node) && node.key.type === AST_NODE_TYPES.Identifier && node.key.name.startsWith("render") && isOneOf([AST_NODE_TYPES.ClassDeclaration, AST_NODE_TYPES.ClassExpression])(node.parent.parent);
}
function isRenderMethodCallback(node) {
	const parent = node.parent;
	const greatGrandparent = parent.parent?.parent;
	return greatGrandparent != null && isRenderMethodLike(parent) && isClassComponent(greatGrandparent);
}
/**
* @param node The call expression node to check.
* @deprecated Class components are legacy. This function exists only to support legacy rules.
*/
function isThisSetStateCall(node) {
	const callee = Extract.unwrap(node.callee);
	return callee.type === AST_NODE_TYPES.MemberExpression && callee.object.type === AST_NODE_TYPES.ThisExpression && callee.property.type === AST_NODE_TYPES.Identifier && callee.property.name === "setState";
}
/**
* @param node The assignment expression node to check.
* @deprecated Class components are legacy. This function exists only to support legacy rules.
*/
function isAssignmentToThisState(node) {
	const { left } = node;
	let current = Extract.unwrap(left);
	while (current.type === AST_NODE_TYPES.MemberExpression) {
		const { object, property } = current;
		if (object.type === AST_NODE_TYPES.ThisExpression && Extract.getPropertyName(property) === "state") return true;
		current = Extract.unwrap(object);
	}
	return false;
}

//#endregion
//#region src/class-component-collector.ts
/**
* @param context The rule context.
* @deprecated Class components are legacy. This function exists only to support legacy rules.
*/
function getClassComponentCollector(context) {
	const components = /* @__PURE__ */ new Map();
	const api = { getAllComponents(_) {
		return [...components.values()];
	} };
	const getText = (n) => context.sourceCode.getText(n);
	const collect = (node) => {
		if (!isClassComponent(node)) return;
		const id = getClassId(node);
		const key = randomBytes(8).toString("hex");
		const name = id == null ? null : Extract.getFullyQualifiedName(id, getText);
		components.set(key, {
			id,
			key,
			kind: "class-component",
			name,
			displayName: null,
			flag: 0n,
			hint: 0n,
			methods: [],
			node
		});
	};
	return {
		api,
		visitor: {
			ClassDeclaration: collect,
			ClassExpression: collect
		}
	};
}

//#endregion
//#region src/function.ts
/**
* Gets the static identifier of a function AST node.
*
* @remarks
* For function declarations this is straightforward. For anonymous function
* expressions it is more complex. This function roughly detects the same AST
* nodes as the ECMAScript spec's `IsAnonymousFunctionDefinition()` with some
* exceptions to better fit our use case.
*
* Ported from {@link https://github.com/facebook/react/blob/bb8a76c6cc77ea2976d690ea09f5a1b3d9b1792a/packages/eslint-plugin-react-hooks/src/rules/RulesOfHooks.ts#L860 | RulesOfHooks.ts}
*
* @param node - The function node to analyze.
* @returns The identifier node if found, `null` otherwise.
*/
function getFunctionId(node) {
	switch (true) {
		case "id" in node && node.id != null: return node.id;
		case node.parent.type === AST_NODE_TYPES.VariableDeclarator && node.parent.init === node: return node.parent.id;
		case node.parent.type === AST_NODE_TYPES.AssignmentExpression && node.parent.right === node && node.parent.operator === "=": return node.parent.left;
		case node.parent.type === AST_NODE_TYPES.Property && node.parent.value === node && !node.parent.computed: return node.parent.key;
		case Check.isPropertyOrMethod(node.parent) && node.parent.value === node: return node.parent.key;
		case node.parent.type === AST_NODE_TYPES.AssignmentPattern && node.parent.right === node: return node.parent.left;
		case node.parent.type === AST_NODE_TYPES.ConditionalExpression: return getFunctionId(node.parent);
		case Check.isTypeAssertionExpression(node.parent): return getFunctionId(node.parent);
	}
	return null;
}
/**
* Identifies the initialization path of a function node in the AST.
*
* @param node - The function node to analyze.
* @returns The function initialization path or `null` if not identifiable.
*/
function getFunctionInitPath(node) {
	if (node.type === AST_NODE_TYPES.FunctionDeclaration) return [node];
	let parent = node.parent;
	while (Check.isTypeExpression(parent)) parent = parent.parent;
	switch (true) {
		case parent.type === AST_NODE_TYPES.VariableDeclarator: return [
			parent.parent,
			parent,
			node
		];
		case parent.type === AST_NODE_TYPES.CallExpression && parent.parent.type === AST_NODE_TYPES.VariableDeclarator: return [
			parent.parent.parent,
			parent.parent,
			parent,
			node
		];
		case parent.type === AST_NODE_TYPES.CallExpression && parent.parent.type === AST_NODE_TYPES.CallExpression && parent.parent.parent.type === AST_NODE_TYPES.VariableDeclarator: return [
			parent.parent.parent.parent,
			parent.parent.parent,
			parent.parent,
			parent,
			node
		];
		case parent.type === AST_NODE_TYPES.Property && parent.parent.type === AST_NODE_TYPES.ObjectExpression && parent.parent.parent.type === AST_NODE_TYPES.VariableDeclarator: return [
			parent.parent.parent.parent,
			parent.parent.parent,
			parent.parent,
			parent,
			node
		];
		case parent.type === AST_NODE_TYPES.MethodDefinition: return [
			parent.parent.parent,
			parent.parent,
			parent,
			node
		];
		case parent.type === AST_NODE_TYPES.PropertyDefinition: return [
			parent.parent.parent,
			parent.parent,
			parent,
			node
		];
	}
	return null;
}
/**
* Checks if a specific function call exists in the function initialization path.
*
* @param callName - The name of the call to check for (e.g., "memo", "forwardRef").
* @param initPath - The function initialization path to search in.
* @returns `true` if the call exists in the path, `false` otherwise.
*/
function isFunctionHasCallInInitPath(callName, initPath) {
	return initPath.some((node) => {
		if (node.type !== AST_NODE_TYPES.CallExpression) return false;
		const callee = Extract.unwrap(node.callee);
		if (callee.type === AST_NODE_TYPES.Identifier) return callee.name === callName;
		if (callee.type === AST_NODE_TYPES.MemberExpression && "name" in callee.property) return callee.property.name === callName;
		return false;
	});
}
/**
* Checks if a function is empty.
*
* @param node - The function node to check.
* @returns `true` if the function is empty, `false` otherwise.
*/
function isFunctionEmpty(node) {
	return node.body.type === AST_NODE_TYPES.BlockStatement && node.body.body.length === 0;
}
function getFunctionDirectives(node) {
	const directives = [];
	if (node.body.type !== AST_NODE_TYPES.BlockStatement) return directives;
	function isDirective(node) {
		return node.type === AST_NODE_TYPES.ExpressionStatement && node.directive != null;
	}
	for (const stmt of node.body.body) {
		if (!isDirective(stmt)) continue;
		directives.push(stmt);
	}
	return directives;
}
/**
* Checks if a directive with the given name exists in the function directives.
*
* @param node - The function AST node.
* @param name - The directive name to check (e.g., "use memo", "use no memo").
* @returns `true` if the directive exists, `false` otherwise.
*/
function isFunctionHasDirective(node, name) {
	return getFunctionDirectives(node).some((d) => d.directive === name);
}
const SEL_FUNCTION_DISPLAY_NAME_ASSIGNMENT = [
	"AssignmentExpression",
	"[operator='=']",
	"[left.type='MemberExpression']",
	"[left.property.name='displayName']"
].join("");

//#endregion
//#region src/function-component.ts
/**
* Component flag constants
*/
const FunctionComponentFlag = {
	/** No flags set */
	None: 0n,
	/** Indicates the component is a pure component (ex: extends PureComponent) */
	PureComponent: 1n << 0n,
	/** Indicates the component creates elements using `createElement` instead of JSX */
	CreateElement: 1n << 1n,
	/** Indicates the component is memoized (ex: React.memo) */
	Memo: 1n << 2n,
	/** Indicates the component forwards a ref (ex: React.forwardRef) */
	ForwardRef: 1n << 3n
};
/**
* Get component flag from init path
* @param initPath The init path of the function component
* @returns The component flag
* @internal
*/
function getFunctionComponentFlagFromInitPath(initPath) {
	let flag = FunctionComponentFlag.None;
	if (initPath != null && isFunctionHasCallInInitPath("memo", initPath)) flag |= FunctionComponentFlag.Memo;
	if (initPath != null && isFunctionHasCallInInitPath("forwardRef", initPath)) flag |= FunctionComponentFlag.ForwardRef;
	return flag;
}
/**
* Check if the node is a call expression for a component wrapper
* @param context The ESLint rule context
* @param node The node to check
* @returns `true` if the node is a call expression for a component wrapper
*/
function isFunctionComponentWrapperCall(context, node) {
	if (node.type !== AST_NODE_TYPES.CallExpression) return false;
	return isMemoCall(context, node) || isForwardRefCall(context, node);
}
/**
* Check if the node is a callback function passed to a component wrapper
* @param context The ESLint rule context
* @param node The node to check
* @returns `true` if the node is a callback function passed to a component wrapper
*/
function isFunctionComponentWrapperCallback(context, node) {
	if (!Check.isFunction(node)) return false;
	let parent = node.parent;
	while (Check.isTypeExpression(parent)) parent = parent.parent;
	if (parent.type !== AST_NODE_TYPES.CallExpression) return false;
	return isFunctionComponentWrapperCall(context, parent);
}
/**
* Get function component identifier from `const Component = memo(() => {});`
* @param context The rule context
* @param node The AST node to get the function component identifier from
* @internal
*/
function getFunctionComponentId(context, node) {
	const functionId = getFunctionId(node);
	if (functionId != null) return functionId;
	let parent = node.parent;
	while (Check.isTypeExpression(parent)) parent = parent.parent;
	if (parent.type === AST_NODE_TYPES.CallExpression && isFunctionComponentWrapperCall(context, parent) && parent.parent.type === AST_NODE_TYPES.VariableDeclarator) return parent.parent.id;
	if (parent.type === AST_NODE_TYPES.CallExpression && isFunctionComponentWrapperCall(context, parent) && parent.parent.type === AST_NODE_TYPES.CallExpression && isFunctionComponentWrapperCall(context, parent.parent) && parent.parent.parent.type === AST_NODE_TYPES.VariableDeclarator) return parent.parent.parent.id;
	return null;
}
/**
* Check if a string matches the strict component name pattern
* @param name The name to check
*/
function isFunctionComponentName(name) {
	return RE_COMPONENT_NAME.test(name);
}
/**
* Check if a string matches the loose component name pattern
* @param name The name to check
*/
function isFunctionComponentNameLoose(name) {
	return RE_COMPONENT_NAME_LOOSE.test(name);
}
/**
* Check if a function has a loose component name
* @param context The rule context
* @param fn The function to check
* @param allowNone Whether to allow no name
* @returns Whether the function has a loose component name
*/
function isFunctionWithLooseComponentName(context, fn, allowNone = false) {
	const id = getFunctionComponentId(context, fn);
	if (id == null) return allowNone;
	if (id.type === AST_NODE_TYPES.Identifier) return isFunctionComponentNameLoose(id.name);
	if (id.type === AST_NODE_TYPES.MemberExpression && id.property.type === AST_NODE_TYPES.Identifier) return isFunctionComponentNameLoose(id.property.name);
	return false;
}
/**
* Hints for component collector
*/
const FunctionComponentDetectionHint = {
	...JsxDetectionHint$1,
	DoNotIncludeFunctionDefinedAsClassMethod: 1n << 11n,
	DoNotIncludeFunctionDefinedAsClassProperty: 1n << 12n,
	DoNotIncludeFunctionDefinedAsObjectMethod: 1n << 13n,
	DoNotIncludeFunctionDefinedAsArrayExpressionElement: 1n << 14n,
	DoNotIncludeFunctionDefinedAsArrayPatternElement: 1n << 15n,
	DoNotIncludeFunctionDefinedAsArbitraryCallExpressionCallback: 1n << 18n,
	DoNotIncludeFunctionDefinedAsArrayFlatMapCallback: 1n << 16n,
	DoNotIncludeFunctionDefinedAsArrayMapCallback: 1n << 17n
};
/**
* Default component detection hint
*/
const DEFAULT_COMPONENT_DETECTION_HINT = 0n | FunctionComponentDetectionHint.DoNotIncludeJsxWithBigIntValue | FunctionComponentDetectionHint.DoNotIncludeJsxWithBooleanValue | FunctionComponentDetectionHint.DoNotIncludeJsxWithNumberValue | FunctionComponentDetectionHint.DoNotIncludeJsxWithStringValue | FunctionComponentDetectionHint.DoNotIncludeJsxWithUndefinedValue | FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArbitraryCallExpressionCallback | FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayExpressionElement | FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayFlatMapCallback | FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayMapCallback | FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayPatternElement | FunctionComponentDetectionHint.RequireAllArrayElementsToBeJsx | FunctionComponentDetectionHint.RequireBothBranchesOfConditionalExpressionToBeJsx | FunctionComponentDetectionHint.RequireBothSidesOfLogicalExpressionToBeJsx;
/**
* Determine if a function node represents a valid React component definition
*
* @param context The rule context
* @param node The function node to analyze
* @param hint Component detection hints (bit flags) to customize detection logic
* @returns `true` if the node is considered a component definition
*/
function isFunctionComponentDefinition(context, node, hint) {
	if (!isFunctionWithLooseComponentName(context, node, true)) return false;
	const isCreateElementArg = (() => {
		let p = node.parent;
		while (Check.isTypeExpression(p)) p = p.parent;
		if (p.type !== AST_NODE_TYPES.CallExpression || !isCreateElementCall(context, p)) return false;
		return p.arguments.slice(2).some((arg) => Extract.unwrap(arg) === node);
	})();
	switch (true) {
		case isCreateElementArg: return false;
		case isRenderMethodCallback(node): return false;
	}
	let parent = node.parent;
	while (Check.isTypeExpression(parent)) parent = parent.parent;
	const parentCallee = parent.type === AST_NODE_TYPES.CallExpression ? Extract.unwrap(parent.callee) : null;
	switch (true) {
		case isOneOf([AST_NODE_TYPES.ArrowFunctionExpression, AST_NODE_TYPES.FunctionExpression])(node) && parent.type === AST_NODE_TYPES.Property && parent.parent.type === AST_NODE_TYPES.ObjectExpression:
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsObjectMethod) return false;
			break;
		case isOneOf([AST_NODE_TYPES.ArrowFunctionExpression, AST_NODE_TYPES.FunctionExpression])(node) && parent.type === AST_NODE_TYPES.MethodDefinition:
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsClassMethod) return false;
			break;
		case isOneOf([AST_NODE_TYPES.ArrowFunctionExpression, AST_NODE_TYPES.FunctionExpression])(node) && parent.type === AST_NODE_TYPES.Property:
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsClassProperty) return false;
			break;
		case parent.type === AST_NODE_TYPES.ArrayPattern:
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayPatternElement) return false;
			break;
		case parent.type === AST_NODE_TYPES.ArrayExpression:
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayExpressionElement) return false;
			break;
		case parentCallee != null && parentCallee.type === AST_NODE_TYPES.MemberExpression && parentCallee.property.type === AST_NODE_TYPES.Identifier && parentCallee.property.name === "map":
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayMapCallback) return false;
			break;
		case parentCallee != null && parentCallee.type === AST_NODE_TYPES.MemberExpression && parentCallee.property.type === AST_NODE_TYPES.Identifier && parentCallee.property.name === "flatMap":
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArrayFlatMapCallback) return false;
			break;
		case parent.type === AST_NODE_TYPES.CallExpression && getFunctionId(node) == null && !isFunctionComponentWrapperCall(context, parent) && !isCreateElementCall(context, parent):
			if (hint & FunctionComponentDetectionHint.DoNotIncludeFunctionDefinedAsArbitraryCallExpressionCallback) return false;
			break;
	}
	const significantParent = Traverse.findParent(node, isOneOf([
		AST_NODE_TYPES.JSXExpressionContainer,
		AST_NODE_TYPES.ArrowFunctionExpression,
		AST_NODE_TYPES.FunctionExpression,
		AST_NODE_TYPES.Property,
		AST_NODE_TYPES.ClassBody
	]));
	if (significantParent == null) return true;
	if (significantParent.type === AST_NODE_TYPES.JSXExpressionContainer) return false;
	return true;
}

//#endregion
//#region src/hook.ts
const REACT_BUILTIN_HOOK_NAMES = [
	"use",
	"useActionState",
	"useCallback",
	"useContext",
	"useDebugValue",
	"useDeferredValue",
	"useEffect",
	"useFormStatus",
	"useId",
	"useImperativeHandle",
	"useInsertionEffect",
	"useLayoutEffect",
	"useMemo",
	"useOptimistic",
	"useReducer",
	"useRef",
	"useState",
	"useSyncExternalStore",
	"useTransition"
];
/**
* Catch all identifiers that begin with "use" followed by an uppercase Latin
* character to exclude identifiers like "user".
* @param name The name of the identifier to check.
* @see https://github.com/facebook/react/blob/1d6c8168db1d82713202e842df3167787ffa00ed/packages/eslint-plugin-react-hooks/src/rules/RulesOfHooks.ts#L16
*/
function isHookName(name) {
	return name === "use" || /^use[A-Z0-9]/.test(name);
}
/**
* Checks if the given node is a hook identifier
* @param id The AST node to check
* @returns `true` if the node is a hook identifier or member expression with hook name, `false` otherwise
*/
function isHookId(id) {
	switch (id.type) {
		case AST_NODE_TYPES.Identifier: return isHookName(id.name);
		case AST_NODE_TYPES.MemberExpression: return "name" in id.property && isHookName(id.property.name);
		default: return false;
	}
}
/**
* Determine if a function node is a React Hook based on its name.
* @param node The function node to check
* @returns True if the function is a React Hook, false otherwise
*/
function isHookDefinition(node) {
	if (node == null) return false;
	const id = getFunctionId(node);
	switch (id?.type) {
		case AST_NODE_TYPES.Identifier: return isHookName(id.name);
		case AST_NODE_TYPES.MemberExpression: return "name" in id.property && isHookName(id.property.name);
		default: return false;
	}
}
/**
* Check if the given node is a React Hook call by its name.
* @param node The node to check.
* @returns `true` if the node is a React Hook call, `false` otherwise.
*/
function isHookCall(node) {
	if (node == null) return false;
	if (node.type !== AST_NODE_TYPES.CallExpression) return false;
	const callee = Extract.unwrap(node.callee);
	if (callee.type === AST_NODE_TYPES.Identifier) return isHookName(callee.name);
	if (callee.type === AST_NODE_TYPES.MemberExpression) return callee.property.type === AST_NODE_TYPES.Identifier && isHookName(callee.property.name);
	return false;
}
/**
* Detect useEffect calls and variations (useLayoutEffect, etc.) using a regex pattern
* @param node The AST node to check
* @param additionalEffectHooks Regex pattern matching custom hooks that should be treated as effect hooks
* @returns True if the node is a useEffect-like call
*/
function isUseEffectLikeCall(node, additionalEffectHooks = { test: constFalse }) {
	if (node == null) return false;
	if (node.type !== AST_NODE_TYPES.CallExpression) return false;
	const callee = Extract.unwrap(node.callee);
	return [/^use\w*Effect$/u, additionalEffectHooks].some((regexp) => {
		if (callee.type === AST_NODE_TYPES.Identifier) return regexp.test(callee.name);
		if (callee.type === AST_NODE_TYPES.MemberExpression) return callee.property.type === AST_NODE_TYPES.Identifier && regexp.test(callee.property.name);
		return false;
	});
}
/**
* Detect useState calls and variations using a regex pattern
* @param node The AST node to check
* @param additionalStateHooks Regex pattern matching custom hooks that should be treated as state hooks
* @returns True if the node is a useState-like call
*/
function isUseStateLikeCall(node, additionalStateHooks = { test: constFalse }) {
	if (node == null) return false;
	if (node.type !== AST_NODE_TYPES.CallExpression) return false;
	const callee = Extract.unwrap(node.callee);
	switch (true) {
		case callee.type === AST_NODE_TYPES.Identifier: return callee.name === "useState" || additionalStateHooks.test(callee.name);
		case callee.type === AST_NODE_TYPES.MemberExpression && callee.property.type === AST_NODE_TYPES.Identifier: return Extract.getPropertyName(callee.property) === "useState" || additionalStateHooks.test(callee.property.name);
	}
	return false;
}
/**
* Determine if a node is the setup function passed to a useEffect-like hook
* @param node The AST node to check
*/
function isUseEffectSetupCallback(node) {
	if (node == null) return false;
	const expr = Extract.unwrap(node);
	return expr.parent?.type === AST_NODE_TYPES.CallExpression && expr.parent.arguments.at(0) === expr && isUseEffectLikeCall(expr.parent);
}
/**
* Determine if a node is the cleanup function returned by a useEffect-like hook's setup function
* @param node The AST node to check
*/
function isUseEffectCleanupCallback(node) {
	if (node == null) return false;
	const expr = Extract.unwrap(node);
	const returnStatement = Traverse.findParent(expr, is(AST_NODE_TYPES.ReturnStatement));
	const enclosingFunction = Traverse.findParent(expr, Check.isFunction);
	if (enclosingFunction !== Traverse.findParent(returnStatement, Check.isFunction)) return false;
	return isUseEffectSetupCallback(enclosingFunction);
}

//#endregion
//#region src/jsx.ts
const JsxDetectionHint = {
	None: 0n,
	DoNotIncludeJsxWithNullValue: 1n << 0n,
	DoNotIncludeJsxWithNumberValue: 1n << 1n,
	DoNotIncludeJsxWithBigIntValue: 1n << 2n,
	DoNotIncludeJsxWithStringValue: 1n << 3n,
	DoNotIncludeJsxWithBooleanValue: 1n << 4n,
	DoNotIncludeJsxWithUndefinedValue: 1n << 5n,
	DoNotIncludeJsxWithEmptyArrayValue: 1n << 6n,
	DoNotIncludeJsxWithCreateElementValue: 1n << 7n,
	RequireAllArrayElementsToBeJsx: 1n << 8n,
	RequireBothSidesOfLogicalExpressionToBeJsx: 1n << 9n,
	RequireBothBranchesOfConditionalExpressionToBeJsx: 1n << 10n
};
/**
* Default JSX detection configuration.
*
* Skips number, bigint, boolean, string, and undefined literals –
* the value types that are commonly returned alongside JSX in React
* components but are not themselves renderable elements.
*/
const DEFAULT_JSX_DETECTION_HINT = 0n | JsxDetectionHint.DoNotIncludeJsxWithNumberValue | JsxDetectionHint.DoNotIncludeJsxWithBigIntValue | JsxDetectionHint.DoNotIncludeJsxWithBooleanValue | JsxDetectionHint.DoNotIncludeJsxWithStringValue | JsxDetectionHint.DoNotIncludeJsxWithUndefinedValue;
/**
* Determine whether a node represents JSX-like content based on heuristics.
*
* The detection behaviour is configurable through {@link JsxDetectionHint}
* bit-flags so that callers can opt individual value kinds in or out.
*
* @param context - The ESLint rule context (needed for variable resolution).
* @param node    - The AST node to analyse.
* @param hint    - Optional bit-flags to adjust detection behaviour.
*                  Defaults to {@link DEFAULT_JSX_DETECTION_HINT}.
* @returns Whether the node is considered JSX-like.
*
* @example
* ```ts
* import { isJsxLike } from "@eslint-react/core";
*
* if (isJsxLike(context, node)) {
*   // node looks like it evaluates to a React element
* }
* ```
*/
function isJsxLike(context, node, hint = DEFAULT_JSX_DETECTION_HINT) {
	if (node == null) return false;
	if (Check.isJSX(node)) return true;
	switch (node.type) {
		case AST_NODE_TYPES.Literal:
			switch (typeof node.value) {
				case "boolean": return !(hint & JsxDetectionHint.DoNotIncludeJsxWithBooleanValue);
				case "string": return !(hint & JsxDetectionHint.DoNotIncludeJsxWithStringValue);
				case "number": return !(hint & JsxDetectionHint.DoNotIncludeJsxWithNumberValue);
				case "bigint": return !(hint & JsxDetectionHint.DoNotIncludeJsxWithBigIntValue);
			}
			if (node.value == null) return !(hint & JsxDetectionHint.DoNotIncludeJsxWithNullValue);
			return false;
		case AST_NODE_TYPES.TemplateLiteral: return !(hint & JsxDetectionHint.DoNotIncludeJsxWithStringValue);
		case AST_NODE_TYPES.ArrayExpression:
			if (node.elements.length === 0) return !(hint & JsxDetectionHint.DoNotIncludeJsxWithEmptyArrayValue);
			if (hint & JsxDetectionHint.RequireAllArrayElementsToBeJsx) return node.elements.every((n) => isJsxLike(context, n, hint));
			return node.elements.some((n) => isJsxLike(context, n, hint));
		case AST_NODE_TYPES.LogicalExpression:
			if (hint & JsxDetectionHint.RequireBothSidesOfLogicalExpressionToBeJsx) return isJsxLike(context, node.left, hint) && isJsxLike(context, node.right, hint);
			return isJsxLike(context, node.left, hint) || isJsxLike(context, node.right, hint);
		case AST_NODE_TYPES.ConditionalExpression: {
			const consequentIsJsx = Array.isArray(node.consequent) ? checkArray(context, node.consequent, hint) : isJsxLike(context, node.consequent, hint);
			const alternateIsJsx = isJsxLike(context, node.alternate, hint);
			if (hint & JsxDetectionHint.RequireBothBranchesOfConditionalExpressionToBeJsx) return consequentIsJsx && alternateIsJsx;
			return consequentIsJsx || alternateIsJsx;
		}
		case AST_NODE_TYPES.SequenceExpression: return isJsxLike(context, node.expressions.at(-1) ?? null, hint);
		case AST_NODE_TYPES.CallExpression: {
			if (hint & JsxDetectionHint.DoNotIncludeJsxWithCreateElementValue) return false;
			const callee = Extract.unwrap(node.callee);
			switch (callee.type) {
				case AST_NODE_TYPES.Identifier: return callee.name === "createElement";
				case AST_NODE_TYPES.MemberExpression: return callee.property.type === AST_NODE_TYPES.Identifier && callee.property.name === "createElement";
			}
			return false;
		}
		case AST_NODE_TYPES.Identifier:
			if (node.name === "undefined") return !(hint & JsxDetectionHint.DoNotIncludeJsxWithUndefinedValue);
			if (Check.isJSXTagNameExpression(node)) return true;
			return isJsxLike(context, resolve(context, node), hint);
	}
	return false;
}
function checkArray(context, elements, hint) {
	if (elements.length === 0) return !(hint & JsxDetectionHint.DoNotIncludeJsxWithEmptyArrayValue);
	if (hint & JsxDetectionHint.RequireAllArrayElementsToBeJsx) return elements.every((n) => isJsxLike(context, n, hint));
	return elements.some((n) => isJsxLike(context, n, hint));
}

//#endregion
//#region src/function-component-collector.ts
/**
* Get an api and visitor object for the rule to collect function components
* @param context The ESLint rule context
* @param options The options to use
* @returns The api and visitor of the collector
*/
function getFunctionComponentCollector(context, options = {}) {
	const { collectDisplayName = false, hint = DEFAULT_COMPONENT_DETECTION_HINT } = options;
	const functionEntries = [];
	const components = /* @__PURE__ */ new Map();
	const getText = (n) => context.sourceCode.getText(n);
	const getCurrentEntry = () => functionEntries.at(-1) ?? null;
	const onFunctionEnter = (node) => {
		const key = randomBytes(8).toString("hex");
		const exp = Traverse.findParent(node, (n) => n.type === AST_NODE_TYPES.ExportDefaultDeclaration);
		const isExportDefault = exp != null;
		const isExportDefaultDeclaration = exp != null && Extract.unwrap(exp.declaration) === node;
		const id = getFunctionComponentId(context, node);
		const name = id == null ? null : Extract.getFullyQualifiedName(id, getText);
		const initPath = getFunctionInitPath(node);
		const directives = getFunctionDirectives(node);
		const entry = {
			id,
			key,
			kind: "component",
			name,
			directives,
			displayName: null,
			flag: getFunctionComponentFlagFromInitPath(initPath),
			hint,
			hookCalls: [],
			initPath,
			isExportDefault,
			isExportDefaultDeclaration,
			isFunctionComponentDefinition: isFunctionComponentDefinition(context, node, hint),
			node,
			rets: []
		};
		functionEntries.push(entry);
		if (!entry.isFunctionComponentDefinition || !isFunctionWithLooseComponentName(context, node, false)) return;
		if (directives.some((d) => d.directive === "use memo" || d.directive === "use no memo")) components.set(entry.key, entry);
	};
	const onFunctionExit = () => {
		return functionEntries.pop();
	};
	return {
		api: { getAllComponents(_) {
			return [...components.values()];
		} },
		visitor: {
			":function": onFunctionEnter,
			":function:exit": onFunctionExit,
			"ArrowFunctionExpression[body.type!='BlockStatement']"() {
				const entry = getCurrentEntry();
				if (entry == null) return;
				const { body } = entry.node;
				if (body.type === AST_NODE_TYPES.BlockStatement) return;
				entry.rets.push(body);
				if (!entry.isFunctionComponentDefinition) return;
				if (!components.has(entry.key) && !isJsxLike(context, body, hint)) return;
				components.set(entry.key, entry);
			},
			...collectDisplayName ? { [SEL_FUNCTION_DISPLAY_NAME_ASSIGNMENT](node) {
				const { left, right } = node;
				if (left.type !== AST_NODE_TYPES.MemberExpression) return;
				const componentName = left.object.type === AST_NODE_TYPES.Identifier ? left.object.name : null;
				const component = [...components.values()].findLast(({ name }) => name != null && name === componentName);
				if (component == null) return;
				component.displayName = right;
			} } : {},
			CallExpression(node) {
				if (!isHookCall(node)) return;
				const entry = getCurrentEntry();
				if (entry == null) return;
				entry.hookCalls.push(node);
				if (!entry.isFunctionComponentDefinition) return;
				components.set(entry.key, entry);
			},
			ReturnStatement(node) {
				const entry = getCurrentEntry();
				if (entry == null) return;
				entry.rets.push(node.argument);
				if (!entry.isFunctionComponentDefinition) return;
				const { argument } = node;
				if (!components.has(entry.key) && !isJsxLike(context, argument, hint)) return;
				components.set(entry.key, entry);
			}
		}
	};
}

//#endregion
//#region src/hook-collector.ts
/**
* Get an api and visitor object for the rule to collect hooks
* @param context The ESLint rule context
* @returns The api and visitor of the collector
*/
function getHookCollector(context) {
	const hooks = /* @__PURE__ */ new Map();
	const functionEntries = [];
	const getText = (n) => context.sourceCode.getText(n);
	const getCurrentEntry = () => functionEntries.at(-1) ?? null;
	const onFunctionEnter = (node) => {
		const id = getFunctionId(node);
		const key = randomBytes(8).toString("hex");
		const entry = {
			id,
			key,
			kind: "hook",
			name: id == null ? null : Extract.getFullyQualifiedName(id, getText),
			directives: [],
			flag: 0n,
			hint: 0n,
			hookCalls: [],
			isHookDefinition: id != null && isHookId(id),
			node,
			rets: []
		};
		functionEntries.push(entry);
		if (!entry.isHookDefinition) return;
		hooks.set(key, entry);
	};
	const onFunctionExit = () => {
		functionEntries.pop();
	};
	return {
		api: { getAllHooks(_) {
			return [...hooks.values()];
		} },
		visitor: {
			":function": onFunctionEnter,
			":function:exit": onFunctionExit,
			"ArrowFunctionExpression[body.type!='BlockStatement']"() {
				const entry = getCurrentEntry();
				if (entry == null) return;
				const { body } = entry.node;
				if (body.type === AST_NODE_TYPES.BlockStatement) return;
				entry.rets.push(body);
			},
			CallExpression(node) {
				if (!isHookCall(node)) return;
				const entry = getCurrentEntry();
				if (entry == null) return;
				entry.hookCalls.push(node);
			},
			ReturnStatement(node) {
				const entry = getCurrentEntry();
				if (entry == null) return;
				entry.rets.push(node.argument);
			}
		}
	};
}

//#endregion
//#region src/jsx-config.ts
/**
* Weak‑map cache keyed by `sourceCode` so that the (potentially expensive)
* pragma‑scanning pass runs at most once per file.
*/
const annotationCache = /* @__PURE__ */ new WeakMap();
/**
* Weak‑map cache for the fully‑merged config (compiler options + annotation).
*/
const mergedCache = /* @__PURE__ */ new WeakMap();
/**
* Read JSX configuration from the TypeScript compiler options exposed by the
* parser services.
*
* Falls back to sensible React defaults when no compiler options are
* available (e.g. when the file is parsed without type information).
*
* @param context - The ESLint rule context.
* @returns Fully‑populated `JsxConfig` derived from compiler options.
*/
function getJsxConfigFromCompilerOptions(context) {
	const options = context.sourceCode.parserServices?.program?.getCompilerOptions() ?? {};
	return {
		jsx: options.jsx ?? ts.JsxEmit.ReactJSX,
		jsxFactory: options.jsxFactory ?? "React.createElement",
		jsxFragmentFactory: options.jsxFragmentFactory ?? "React.Fragment",
		jsxImportSource: options.jsxImportSource ?? "react"
	};
}
/**
* Extract JSX configuration from `@jsx`, `@jsxFrag`, `@jsxRuntime` and
* `@jsxImportSource` pragma comments in the source file.
*
* The result is cached per `sourceCode` instance via a `WeakMap` so that
* repeated calls from different rules analysing the same file are free.
*
* @param context - The ESLint rule context.
* @returns Partial `JsxConfig` containing only the values found in pragmas.
*/
function getJsxConfigFromAnnotation(context) {
	const cached = annotationCache.get(context.sourceCode);
	if (cached != null) return cached;
	const options = {};
	if (!context.sourceCode.text.includes("@jsx")) {
		annotationCache.set(context.sourceCode, options);
		return options;
	}
	let jsx, jsxFrag, jsxRuntime, jsxImportSource;
	for (const comment of context.sourceCode.getAllComments().reverse()) {
		const value = comment.value;
		jsx ??= value.match(RE_ANNOTATION_JSX)?.[1];
		jsxFrag ??= value.match(RE_ANNOTATION_JSX_FRAG)?.[1];
		jsxRuntime ??= value.match(RE_ANNOTATION_JSX_RUNTIME)?.[1];
		jsxImportSource ??= value.match(RE_ANNOTATION_JSX_IMPORT_SOURCE)?.[1];
	}
	if (jsx != null) options.jsxFactory = jsx;
	if (jsxFrag != null) options.jsxFragmentFactory = jsxFrag;
	if (jsxRuntime != null) options.jsx = jsxRuntime === "classic" ? ts.JsxEmit.React : ts.JsxEmit.ReactJSX;
	if (jsxImportSource != null) options.jsxImportSource = jsxImportSource;
	annotationCache.set(context.sourceCode, options);
	return options;
}
/**
* Get the fully‑merged JSX configuration for the current file.
*
* Compiler options provide the base values; pragma annotations found in the
* source override them where present.  The result is cached per `sourceCode`.
*
* This is the main entry‑point most consumers should use.
*
* @param context - The ESLint rule context.
* @returns Fully‑populated, merged `JsxConfig`.
*/
function getJsxConfig(context) {
	const cached = mergedCache.get(context.sourceCode);
	if (cached != null) return cached;
	const merged = {
		...getJsxConfigFromCompilerOptions(context),
		...getJsxConfigFromAnnotation(context)
	};
	mergedCache.set(context.sourceCode, merged);
	return merged;
}

//#endregion
//#region src/type-is.ts
function isFlagSet(allFlags, flag) {
	return (allFlags & flag) !== 0;
}
function isFlagSetOnObject(obj, flag) {
	return isFlagSet(obj.flags, flag);
}
const isTypeFlagSet = isFlagSetOnObject;
function isBooleanLiteralType(type) {
	return isTypeFlagSet(type, ts.TypeFlags.BooleanLiteral);
}
/** @internal */
const isFalseLiteralType = (type) => isBooleanLiteralType(type) && type.intrinsicName === "false";
/** @internal */
const isTrueLiteralType = (type) => isBooleanLiteralType(type) && type.intrinsicName === "true";
/** @internal */
const isAnyType = (type) => isTypeFlagSet(type, ts.TypeFlags.TypeParameter | ts.TypeFlags.Any);
/** @internal */
const isBigIntType = (type) => isTypeFlagSet(type, ts.TypeFlags.BigIntLike);
/** @internal */
const isBooleanType = (type) => isTypeFlagSet(type, ts.TypeFlags.BooleanLike);
/** @internal */
const isEnumType = (type) => isTypeFlagSet(type, ts.TypeFlags.EnumLike);
/** @internal */
const isFalsyBigIntType = (type) => type.isLiteral() && isMatching({ value: { base10Value: "0" } }, type);
/** @internal */
const isFalsyNumberType = (type) => type.isNumberLiteral() && type.value === 0;
/** @internal */
const isFalsyStringType = (type) => type.isStringLiteral() && type.value === "";
/** @internal */
const isNeverType = (type) => isTypeFlagSet(type, ts.TypeFlags.Never);
/** @internal */
const isNullishType = (type) => isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined | ts.TypeFlags.VoidLike);
/** @internal */
const isNumberType = (type) => isTypeFlagSet(type, ts.TypeFlags.NumberLike);
/** @internal */
const isObjectType = (type) => !isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined | ts.TypeFlags.VoidLike | ts.TypeFlags.BooleanLike | ts.TypeFlags.StringLike | ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike | ts.TypeFlags.TypeParameter | ts.TypeFlags.Any | ts.TypeFlags.Unknown | ts.TypeFlags.Never);
/** @internal */
const isStringType = (type) => isTypeFlagSet(type, ts.TypeFlags.StringLike);
/** @internal */
const isTruthyBigIntType = (type) => type.isLiteral() && isMatching({ value: { base10Value: P.not("0") } }, type);
/** @internal */
const isTruthyNumberType = (type) => type.isNumberLiteral() && type.value !== 0;
/** @internal */
const isTruthyStringType = (type) => type.isStringLiteral() && type.value !== "";
/** @internal */
const isUnknownType = (type) => isTypeFlagSet(type, ts.TypeFlags.Unknown);

//#endregion
//#region src/type-name.ts
/**
* An enhanced version of getFullyQualifiedName that handles cases that original function does not handle
* @param checker TypeScript type checker
* @param symbol Symbol to get fully qualified name for
* @returns Fully qualified name of the symbol
*/
function getFullyQualifiedNameEx(checker, symbol) {
	let name = symbol.name;
	let parent = symbol.declarations?.at(0)?.parent;
	if (parent == null) return checker.getFullyQualifiedName(symbol);
	while (parent.kind !== ts.SyntaxKind.SourceFile) {
		switch (true) {
			case ts.isInterfaceDeclaration(parent):
			case ts.isTypeAliasDeclaration(parent):
			case ts.isEnumDeclaration(parent):
			case ts.isModuleDeclaration(parent):
			case ts.isNamespaceImport(parent):
			case ts.isNamespaceExport(parent):
			case ts.isNamespaceExportDeclaration(parent):
				name = `${parent.name.text}.${name}`;
				break;
			case ts.isPropertySignature(parent) && ts.isIdentifier(parent.name):
			case ts.isPropertyDeclaration(parent) && ts.isIdentifier(parent.name):
			case ts.isMethodDeclaration(parent) && ts.isIdentifier(parent.name):
			case ts.isMethodSignature(parent) && ts.isIdentifier(parent.name):
			case ts.isPropertyAssignment(parent) && ts.isIdentifier(parent.name):
				name = `${parent.name.text}.${name}`;
				break;
			case ts.isFunctionDeclaration(parent) && parent.name != null:
			case ts.isClassExpression(parent) && parent.name != null:
			case ts.isClassDeclaration(parent) && parent.name != null:
				name = `${parent.name.text}.${name}`;
				break;
			case ts.isEnumMember(parent):
				name = `${parent.name.getText()}.${name}`;
				break;
			case ts.isTypeLiteralNode(parent):
			case ts.isMappedTypeNode(parent):
			case ts.isObjectLiteralExpression(parent):
			case ts.isIntersectionTypeNode(parent):
			case ts.isUnionTypeNode(parent): break;
			default: break;
		}
		parent = parent.parent;
	}
	const namespace = parent.getSourceFile().statements.find((n) => ts.isNamespaceExportDeclaration(n))?.name.text;
	if (namespace == null) return name;
	if (name.startsWith(`${namespace}.`)) return name;
	return `${namespace}.${name}`;
}

//#endregion
//#region src/type-variant.ts
/**
* Ported from https://github.com/typescript-eslint/typescript-eslint/blob/eb736bbfc22554694400e6a4f97051d845d32e0b/packages/eslint-plugin/src/rules/strict-boolean-expressions.ts#L826 with some enhancements
* Get the variants of an array of types.
* @param types The types to get the variants of
* @returns The variants of the types
* @internal
*/
function getTypeVariants(types) {
	const variants = /* @__PURE__ */ new Set();
	if (types.some(isUnknownType)) {
		variants.add("unknown");
		return variants;
	}
	if (types.some(isNullishType)) variants.add("nullish");
	const booleans = types.filter(isBooleanType);
	const boolean0 = booleans[0];
	if (booleans.length === 1 && boolean0 != null) {
		if (isFalseLiteralType(boolean0)) variants.add("falsy boolean");
		else if (isTrueLiteralType(boolean0)) variants.add("truthy boolean");
	} else if (booleans.length === 2) variants.add("boolean");
	const strings = types.filter(isStringType);
	if (strings.length > 0) {
		const evaluated = match(strings).when((types) => types.every(isTruthyStringType), () => "truthy string").when((types) => types.every(isFalsyStringType), () => "falsy string").otherwise(() => "string");
		variants.add(evaluated);
	}
	const bigints = types.filter(isBigIntType);
	if (bigints.length > 0) {
		const evaluated = match(bigints).when((types) => types.every(isTruthyBigIntType), () => "truthy bigint").when((types) => types.every(isFalsyBigIntType), () => "falsy bigint").otherwise(() => "bigint");
		variants.add(evaluated);
	}
	const numbers = types.filter(isNumberType);
	if (numbers.length > 0) {
		const evaluated = match(numbers).when((types) => types.every(isTruthyNumberType), () => "truthy number").when((types) => types.every(isFalsyNumberType), () => "falsy number").otherwise(() => "number");
		variants.add(evaluated);
	}
	if (types.some(isEnumType)) variants.add("enum");
	if (types.some(isObjectType)) variants.add("object");
	if (types.some(isAnyType)) variants.add("any");
	if (types.some(isNeverType)) variants.add("never");
	return variants;
}

//#endregion
export { DEFAULT_COMPONENT_DETECTION_HINT, DEFAULT_JSX_DETECTION_HINT, FunctionComponentDetectionHint, FunctionComponentFlag, JsxDetectionHint, REACT_BUILTIN_HOOK_NAMES, SEL_FUNCTION_DISPLAY_NAME_ASSIGNMENT, getClassComponentCollector, getClassId, getFullyQualifiedNameEx, getFunctionComponentCollector, getFunctionComponentFlagFromInitPath, getFunctionComponentId, getFunctionDirectives, getFunctionId, getFunctionInitPath, getHookCollector, getJsxConfig, getJsxConfigFromAnnotation, getJsxConfigFromCompilerOptions, getTypeVariants, isAPI, isAPICall, isAPIFromReact, isAPIFromReactNative, isAnyType, isAssignmentToThisState, isBigIntType, isBooleanLiteralType, isBooleanType, isCaptureOwnerStack, isCaptureOwnerStackCall, isChildrenCount, isChildrenCountCall, isChildrenForEach, isChildrenForEachCall, isChildrenMap, isChildrenMapCall, isChildrenOnly, isChildrenOnlyCall, isChildrenToArray, isChildrenToArrayCall, isClassComponent, isCloneElement, isCloneElementCall, isComponentDidCatch, isComponentDidMount, isComponentDidUpdate, isComponentWillMount, isComponentWillReceiveProps, isComponentWillUnmount, isComponentWillUpdate, isCreateContext, isCreateContextCall, isCreateElement, isCreateElementCall, isCreateRef, isCreateRefCall, isEnumType, isFalseLiteralType, isFalsyBigIntType, isFalsyNumberType, isFalsyStringType, isForwardRef, isForwardRefCall, isFunctionComponentDefinition, isFunctionComponentName, isFunctionComponentNameLoose, isFunctionComponentWrapperCall, isFunctionComponentWrapperCallback, isFunctionEmpty, isFunctionHasCallInInitPath, isFunctionHasDirective, isFunctionWithLooseComponentName, isGetChildContext, isGetDefaultProps, isGetDerivedStateFromError, isGetDerivedStateFromProps, isGetInitialState, isGetSnapshotBeforeUpdate, isHookCall, isHookDefinition, isHookId, isHookName, isJsxLike, isLazy, isLazyCall, isMemo, isMemoCall, isNeverType, isNullishType, isNumberType, isObjectType, isPureComponent, isRender, isRenderMethodCallback, isRenderMethodLike, isShouldComponentUpdate, isStringType, isThisSetStateCall, isTrueLiteralType, isTruthyBigIntType, isTruthyNumberType, isTruthyStringType, isUnknownType, isUnsafeComponentWillMount, isUnsafeComponentWillReceiveProps, isUnsafeComponentWillUpdate, isUse, isUseActionState, isUseActionStateCall, isUseCall, isUseCallback, isUseCallbackCall, isUseContext, isUseContextCall, isUseDebugValue, isUseDebugValueCall, isUseDeferredValue, isUseDeferredValueCall, isUseEffect, isUseEffectCall, isUseEffectCleanupCallback, isUseEffectLikeCall, isUseEffectSetupCallback, isUseFormStatus, isUseFormStatusCall, isUseId, isUseIdCall, isUseImperativeHandle, isUseImperativeHandleCall, isUseInsertionEffect, isUseInsertionEffectCall, isUseLayoutEffect, isUseLayoutEffectCall, isUseMemo, isUseMemoCall, isUseOptimistic, isUseOptimisticCall, isUseReducer, isUseReducerCall, isUseRef, isUseRefCall, isUseState, isUseStateCall, isUseStateLikeCall, isUseSyncExternalStore, isUseSyncExternalStoreCall, isUseTransition, isUseTransitionCall };
