// eslint-plugin-react-jsx/dist/index.js
import { DEFAULT_ESLINT_REACT_SETTINGS } from "@eslint-react/shared";
import { ESLintUtils } from "@typescript-eslint/utils";
import * as core from "@eslint-react/core";
import { merge } from "@eslint-react/eslint";
import { findAttribute, getChildren, getElementFullType, hasAnyAttribute, hasChildren, isFragmentElement, isHostElement, isWhitespaceText } from "@eslint-react/jsx";
import { AST_NODE_TYPES } from "@typescript-eslint/types";
import { Check, Extract, isOneOf } from "@eslint-react/ast";
import ts from "typescript";

//#region \0rolldown/runtime.js
var __defProp = Object.defineProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) {
		__defProp(target, name, {
			get: all[name],
			enumerable: true
		});
	}
	if (!no_symbols) {
		__defProp(target, Symbol.toStringTag, { value: "Module" });
	}
	return target;
};

//#endregion
//#region package.json
var name$2 = "eslint-plugin-react-jsx";
var version = "5.8.10";

//#endregion
//#region src/utils/create-rule.ts
function getDocsUrl(ruleName) {
	return `https://eslint-react.xyz/docs/rules/jsx-${ruleName}`;
}
const createRule = ESLintUtils.RuleCreator(getDocsUrl);

//#endregion
//#region src/rules/no-children-prop-with-children/lib.ts
/**
* Find a `children` property inside an ObjectExpression.
* @param node The object expression.
* @returns The Property node whose key is "children", or null.
*/
function findChildrenProperty$1(node) {
	for (const prop of node.properties) {
		if (prop.type !== AST_NODE_TYPES.Property) continue;
		const key = Extract.unwrap(prop.key);
		if (key.type === AST_NODE_TYPES.Identifier && key.name === "children") return prop;
		if (key.type === AST_NODE_TYPES.Literal && key.value === "children") return prop;
	}
	return null;
}
/**
* Compute the removal range for a JSX attribute, consuming any leading
* whitespace (spaces, tabs, newlines) so the resulting markup stays clean.
* @param context The rule context.
* @param prop The JSX attribute.
*/
function getPropRemovalRange$1(context, prop) {
	const { sourceCode } = context;
	let start = prop.range[0];
	const end = prop.range[1];
	while (start > 0 && /\s/.test(sourceCode.text[start - 1])) start--;
	return [start, end];
}
/**
* Compute the range covering **all** children content of a JSX element or
* fragment (from the start of the first child to the end of the last child).
*
* Returns `null` when there are no children at all.
* @param node The JSX element or fragment.
*/
function getChildrenContentRange(node) {
	if (node.children.length === 0) return null;
	const first = node.children[0];
	const last = node.children[node.children.length - 1];
	return [first.range[0], last.range[1]];
}

//#endregion
//#region src/rules/no-children-prop-with-children/no-children-prop-with-children.ts
const RULE_NAME$7 = "no-children-prop-with-children";
var no_children_prop_with_children_default = createRule({
	meta: {
		type: "problem",
		docs: { description: "Disallows passing 'children' as a prop when children are also passed as nested content." },
		fixable: "code",
		hasSuggestions: true,
		messages: {
			default: "Do not pass 'children' as a prop when the element already has children content.",
			removeChildrenContent: "Remove the nested children content.",
			removeChildrenProp: "Remove the 'children' prop."
		},
		schema: []
	},
	name: RULE_NAME$7,
	create: create$7,
	defaultOptions: []
});
function create$7(context) {
	return merge({
		CallExpression(node) {
			if (!core.isCreateElementCall(context, node)) return;
			const [, propsArg, firstExtra] = node.arguments;
			if (propsArg == null || propsArg.type !== AST_NODE_TYPES.ObjectExpression) return;
			const childrenProp = findChildrenProperty$1(propsArg);
			if (childrenProp == null) return;
			if (firstExtra == null) return;
			context.report({
				messageId: "default",
				node: childrenProp
			});
		},
		JSXElement(node) {
			const childrenProp = findAttribute(context, node, "children");
			if (childrenProp == null) return;
			if (!hasChildren(node)) return;
			if (childrenProp.type !== AST_NODE_TYPES.JSXAttribute) {
				context.report({
					messageId: "default",
					node: childrenProp
				});
				return;
			}
			context.report({
				messageId: "default",
				node: childrenProp,
				suggest: [{
					fix(fixer) {
						const [start, end] = getPropRemovalRange$1(context, childrenProp);
						return fixer.removeRange([start, end]);
					},
					messageId: "removeChildrenProp"
				}, {
					fix(fixer) {
						const range = getChildrenContentRange(node);
						if (range == null) return [];
						return fixer.removeRange(range);
					},
					messageId: "removeChildrenContent"
				}]
			});
		}
	});
}

//#endregion
//#region src/rules/no-children-prop/lib.ts
/**
* Find a `children` property inside an ObjectExpression.
* @param node The object expression.
* @returns The Property node whose key is "children", or null.
*/
function findChildrenProperty(node) {
	for (const prop of node.properties) {
		if (prop.type !== AST_NODE_TYPES.Property) continue;
		const key = Extract.unwrap(prop.key);
		if (key.type === AST_NODE_TYPES.Identifier && key.name === "children") return prop;
		if (key.type === AST_NODE_TYPES.Literal && key.value === "children") return prop;
	}
	return null;
}
/**
* Compute the removal range for a JSX attribute, consuming any leading
* whitespace (spaces, tabs, newlines) so the resulting markup stays clean.
* @param context The rule context.
* @param prop The JSX attribute.
*/
function getPropRemovalRange(context, prop) {
	const { sourceCode } = context;
	let start = prop.range[0];
	const end = prop.range[1];
	while (start > 0 && /\s/.test(sourceCode.text[start - 1])) start--;
	return [start, end];
}
/**
* Extract the text to use as JSX children content from a `children` prop.
*
* - `children="text"`        -> `text`  (raw string, no quotes)
* - `children={<div />}`     -> `<div />`  (JSX element, no braces)
* - `children={<>…</>}`      -> `<>…</>`  (JSX fragment, no braces)
* - `children={expression}`  -> `{expression}`  (wrapped in braces)
* - `children`               -> `null`  (boolean shorthand, cannot extract)
* @param context The rule context.
* @param prop The JSX attribute.
*/
function getChildrenPropText(context, prop) {
	const { sourceCode } = context;
	const { value } = prop;
	if (value == null) return null;
	if (value.type === AST_NODE_TYPES.Literal) return String(value.value);
	if (value.type === AST_NODE_TYPES.JSXExpressionContainer) {
		const { expression } = value;
		if (expression.type === AST_NODE_TYPES.JSXEmptyExpression) return null;
		const exprText = sourceCode.getText(expression);
		if (Check.isJSXElementOrFragment(expression)) return exprText;
		return `{${exprText}}`;
	}
	return null;
}

//#endregion
//#region src/rules/no-children-prop/no-children-prop.ts
const RULE_NAME$6 = "no-children-prop";
var no_children_prop_default = createRule({
	meta: {
		type: "suggestion",
		docs: { description: "Disallows passing 'children' as a prop." },
		fixable: "code",
		hasSuggestions: true,
		messages: {
			default: "Do not pass 'children' as props.",
			moveChildrenToContent: "Move 'children' to element content."
		},
		schema: []
	},
	name: RULE_NAME$6,
	create: create$6,
	defaultOptions: []
});
function create$6(context) {
	return merge({
		CallExpression(node) {
			if (!core.isCreateElementCall(context, node)) return;
			const [, propsArg] = node.arguments;
			if (propsArg == null || propsArg.type !== AST_NODE_TYPES.ObjectExpression) return;
			const childrenProp = findChildrenProperty(propsArg);
			if (childrenProp == null) return;
			context.report({
				messageId: "default",
				node: childrenProp
			});
		},
		JSXElement(node) {
			const childrenProp = findAttribute(context, node, "children");
			if (childrenProp == null) return;
			if (childrenProp.type !== AST_NODE_TYPES.JSXAttribute) {
				context.report({
					messageId: "default",
					node: childrenProp
				});
				return;
			}
			const childrenText = getChildrenPropText(context, childrenProp);
			if (childrenText == null) {
				context.report({
					messageId: "default",
					node: childrenProp
				});
				return;
			}
			context.report({
				messageId: "default",
				node: childrenProp,
				suggest: [{
					fix(fixer) {
						const sourceCode = context.sourceCode;
						const { openingElement } = node;
						const [removeStart, removeEnd] = getPropRemovalRange(context, childrenProp);
						if (openingElement.selfClosing) {
							const tagName = sourceCode.getText(openingElement.name);
							const selfCloseOffset = sourceCode.getText(openingElement).lastIndexOf("/>");
							let wsStart = openingElement.range[0] + selfCloseOffset;
							while (wsStart > removeEnd && /\s/.test(sourceCode.text[wsStart - 1])) wsStart--;
							return [fixer.removeRange([removeStart, removeEnd]), fixer.replaceTextRange([wsStart, openingElement.range[1]], `>${childrenText}</${tagName}>`)];
						}
						const fixes = [fixer.removeRange([removeStart, removeEnd])];
						if (node.closingElement != null) fixes.push(fixer.insertTextBefore(node.closingElement, childrenText));
						return fixes;
					},
					messageId: "moveChildrenToContent"
				}]
			});
		}
	});
}

//#endregion
//#region src/rules/no-comment-textnodes/no-comment-textnodes.ts
const RULE_NAME$5 = "no-comment-textnodes";
var no_comment_textnodes_default = createRule({
	meta: {
		type: "problem",
		docs: { description: "Prevents comment strings from being accidentally inserted into a JSX element's text nodes." },
		messages: { default: "Possible misused comment in text node. Comments inside children section of tag should be placed inside braces." },
		schema: []
	},
	name: RULE_NAME$5,
	create: create$5,
	defaultOptions: []
});
function create$5(context) {
	function hasCommentLike(node) {
		if (isOneOf([AST_NODE_TYPES.JSXAttribute, AST_NODE_TYPES.JSXExpressionContainer])(node.parent)) return false;
		return /^\s*\/(?:\/|\*)/mu.test(context.sourceCode.getText(node));
	}
	const visitorFunction = (node) => {
		if (!Check.isJSXElementOrFragment(node.parent)) return;
		if (!hasCommentLike(node)) return;
		context.report({
			messageId: "default",
			node
		});
	};
	return merge({
		JSXText: visitorFunction,
		Literal: visitorFunction
	});
}

//#endregion
//#region src/rules/no-key-after-spread/no-key-after-spread.ts
const RULE_NAME$4 = "no-key-after-spread";
var no_key_after_spread_default = createRule({
	meta: {
		type: "problem",
		docs: { description: "Prevent patterns that cause deoptimization when using the automatic JSX runtime." },
		messages: { noKeyAfterSpread: "Placing 'key' after spread props causes deoptimization when using the automatic JSX runtime. Put 'key' before any spread props." },
		schema: []
	},
	name: RULE_NAME$4,
	create: create$4,
	defaultOptions: []
});
function create$4(context) {
	const { jsx } = core.getJsxConfig(context);
	if (!(jsx === ts.JsxEmit.ReactJSX || jsx === ts.JsxEmit.ReactJSXDev)) return {};
	return merge({ JSXOpeningElement(node) {
		let firstSpreadPropIndex = null;
		for (const [index, prop] of node.attributes.entries()) {
			if (prop.type === AST_NODE_TYPES.JSXSpreadAttribute) {
				firstSpreadPropIndex ??= index;
				continue;
			}
			if (firstSpreadPropIndex == null) continue;
			if (prop.name.type === AST_NODE_TYPES.JSXIdentifier && prop.name.name === "key" && index > firstSpreadPropIndex) context.report({
				messageId: "noKeyAfterSpread",
				node: prop
			});
		}
	} });
}

//#endregion
//#region src/rules/no-leaked-dollar/no-leaked-dollar.ts
const RULE_NAME$3 = "no-leaked-dollar";
var no_leaked_dollar_default = createRule({
	meta: {
		type: "suggestion",
		docs: { description: "Catches `$` before `{expr}` in JSX — typically from template literal `${expr}` being copy-pasted into JSX without removing the `$`. The `$` \"leaks\" into the rendered output." },
		fixable: "code",
		hasSuggestions: true,
		messages: {
			default: "Leaked '$' in JSX. This '$' will be rendered as text nodes.",
			removeDollarSign: "Remove the text node '$'."
		},
		schema: []
	},
	name: RULE_NAME$3,
	create: create$3,
	defaultOptions: []
});
function create$3(context) {
	/**
	* Visitor function for JSXElement and JSXFragment nodes
	* @param node The JSXElement or JSXFragment node to be checked
	*/
	const visitorFunction = (node) => {
		for (const [index, child] of node.children.entries()) {
			if (child.type !== AST_NODE_TYPES.JSXText || !child.value.endsWith("$")) continue;
			if (node.children[index + 1]?.type !== AST_NODE_TYPES.JSXExpressionContainer) continue;
			if (child.value === "$" && node.children.length === 2) continue;
			const pos = child.loc.end;
			context.report({
				loc: {
					end: {
						column: pos.column,
						line: pos.line
					},
					start: {
						column: pos.column - 1,
						line: pos.line
					}
				},
				messageId: "default",
				node: child,
				suggest: [{
					fix(fixer) {
						return fixer.removeRange([child.range[1] - 1, child.range[1]]);
					},
					messageId: "removeDollarSign"
				}]
			});
		}
	};
	return merge({
		JSXElement: visitorFunction,
		JSXFragment: visitorFunction
	});
}

//#endregion
//#region src/rules/no-leaked-semicolon/no-leaked-semicolon.ts
const RULE_NAME$2 = "no-leaked-semicolon";
var no_leaked_semicolon_default = createRule({
	meta: {
		type: "problem",
		docs: { description: "Catches `;` at the start of JSX text nodes — typically from accidentally placing a statement-ending `;` inside JSX. The `;` \"leaks\" into the rendered output." },
		hasSuggestions: true,
		messages: {
			default: "Leaked ';' in JSX. This ';' will be rendered as text nodes.",
			removeSemicolon: "Remove the text node ';'."
		},
		schema: []
	},
	name: RULE_NAME$2,
	create: create$2,
	defaultOptions: []
});
function hasLeakedSemicolon(text) {
	return text.startsWith(";\n") || text.startsWith(";\r");
}
function create$2(context) {
	const visitorFunction = (node) => {
		if (!Check.isJSXElementOrFragment(node.parent)) return;
		if (!hasLeakedSemicolon(context.sourceCode.getText(node))) return;
		context.report({
			messageId: "default",
			node,
			suggest: [{
				fix(fixer) {
					return fixer.removeRange([node.range[0], node.range[0] + 1]);
				},
				messageId: "removeSemicolon"
			}]
		});
	};
	return merge({
		JSXText: visitorFunction,
		Literal: visitorFunction
	});
}

//#endregion
//#region src/rules/no-namespace/no-namespace.ts
const RULE_NAME$1 = "no-namespace";
var no_namespace_default = createRule({
	meta: {
		type: "problem",
		docs: { description: "Disallow JSX namespace syntax, as React does not support them." },
		messages: { noNamespace: "A React component '{{name}}' must not be in a namespace, as React does not support them." },
		schema: []
	},
	name: RULE_NAME$1,
	create: create$1,
	defaultOptions: []
});
function create$1(context) {
	return merge({ JSXElement(node) {
		const name = getElementFullType(node);
		if (typeof name !== "string" || !name.includes(":")) return;
		context.report({
			data: { name },
			messageId: "noNamespace",
			node: node.openingElement.name
		});
	} });
}

//#endregion
//#region src/rules/no-useless-fragment/lib.ts
/**
* Trim leading / trailing whitespace that spans across a newline.
*
* This is used by the auto-fixer to remove indentation / formatting artefacts
* from the raw source text between a fragment's opening and closing tags,
* producing cleaner unwrapped output. It does **not** model React's rendering
* behaviour — React DOM preserves text nodes exactly as written.
*
* Whitespace that contains a newline is stripped entirely;
* whitespace that stays on the same line is preserved.
* @param text The raw source text between JSX tags.
*/
function trimLikeReact(text) {
	const leadingSpaces = /^\s*/.exec(text)?.[0] ?? "";
	const trailingSpaces = /\s*$/.exec(text)?.[0] ?? "";
	const start = leadingSpaces.includes("\n") ? leadingSpaces.length : 0;
	const end = trailingSpaces.includes("\n") ? text.length - trailingSpaces.length : text.length;
	return text.slice(start, end);
}
/**
* Extract the raw source text of an element's / fragment's children
* (everything between the opening and closing tags).
*
* Returns `""` for self-closing elements like `<Fragment />`.
* @param context The rule context.
* @param node The JSX element or fragment.
*/
function getChildrenSourceText(context, node) {
	const { sourceCode } = context;
	const opener = node.type === AST_NODE_TYPES.JSXFragment ? node.openingFragment : node.openingElement;
	const closer = node.type === AST_NODE_TYPES.JSXFragment ? node.closingFragment : node.closingElement;
	if (opener.type === AST_NODE_TYPES.JSXOpeningElement && opener.selfClosing) return "";
	if (closer == null) return "";
	return sourceCode.text.slice(opener.range[1], closer.range[0]);
}

//#endregion
//#region src/rules/no-useless-fragment/no-useless-fragment.ts
const RULE_NAME = "no-useless-fragment";
const defaultOptions = [{
	allowEmptyFragment: false,
	allowExpressions: true
}];
const schema = [{
	type: "object",
	additionalProperties: false,
	properties: {
		allowEmptyFragment: {
			type: "boolean",
			description: "Allow empty fragments"
		},
		allowExpressions: {
			type: "boolean",
			description: "Allow fragments with a single expression child"
		}
	}
}];
var no_useless_fragment_default = createRule({
	meta: {
		type: "suggestion",
		defaultOptions: [...defaultOptions],
		docs: { description: "Disallows useless fragment elements." },
		fixable: "code",
		messages: { default: "A fragment {{reason}} is useless." },
		schema
	},
	name: RULE_NAME,
	create,
	defaultOptions
});
function create(context, [option]) {
	const { allowEmptyFragment = false, allowExpressions = true } = option;
	const jsxConfig = core.getJsxConfig(context);
	/**
	* Whether the fragment has too few meaningful children to justify its
	* existence (the "contains less than two children" reason).
	* @param node The fragment node to check.
	*/
	function isContentUseless(node) {
		if (node.children.length === 0) return !allowEmptyFragment;
		const insideJsx = Check.isJSXElementOrFragment(node.parent);
		if (!allowExpressions) {
			if (insideJsx) return true;
			if (node.children.length === 1) return true;
		}
		if (allowExpressions && !insideJsx && node.children.length === 1) {
			const child = node.children[0];
			if (child != null && child.type === AST_NODE_TYPES.JSXText) return false;
		}
		const meaningful = getChildren(node);
		if (meaningful.length === 0) return true;
		if (meaningful.length === 1 && meaningful[0].type !== AST_NODE_TYPES.JSXExpressionContainer) return true;
		return false;
	}
	/**
	* Whether it is safe to auto-fix the fragment by unwrapping it.
	* @param node The fragment node to check.
	*/
	function isSafeToFix(node) {
		if (Check.isJSXElementOrFragment(node.parent)) return isHostElement(node.parent);
		if (node.children.length === 0) return false;
		return !node.children.some((child) => {
			if (child.type === AST_NODE_TYPES.JSXExpressionContainer) return true;
			if (child.type === AST_NODE_TYPES.JSXText) return !isWhitespaceText(child);
			return false;
		});
	}
	/**
	* Build an autofix that unwraps the fragment, replacing it with its
	* trimmed children text.  Returns `null` when the fix is unsafe.
	* @param node The fragment node to fix.
	*/
	function buildFix(node) {
		if (!isSafeToFix(node)) return null;
		return (fixer) => {
			const childrenText = getChildrenSourceText(context, node);
			return fixer.replaceText(node, trimLikeReact(childrenText));
		};
	}
	/**
	* Inspect a fragment node and report if it is useless.
	*
	* A fragment may be reported for **two independent reasons** on the same
	* node (e.g. `<p><>foo</></p>` is both "placed inside a host component"
	* and* "contains less than two children").
	* @param node The fragment node to inspect.
	*/
	function checkNode(node) {
		if (isHostElement(node.parent)) context.report({
			data: { reason: "placed inside a host component" },
			fix: buildFix(node),
			messageId: "default",
			node
		});
		if (isContentUseless(node)) context.report({
			data: { reason: "contains less than two children" },
			fix: buildFix(node),
			messageId: "default",
			node
		});
	}
	return merge({
		JSXElement(node) {
			if (!isFragmentElement(node, jsxConfig.jsxFragmentFactory)) return;
			if (hasAnyAttribute(context, node, ["key", "ref"])) return;
			checkNode(node);
		},
		JSXFragment(node) {
			checkNode(node);
		}
	});
}

//#endregion
//#region src/plugin.ts
const plugin = {
	meta: {
		name: name$2,
		version
	},
	rules: {
		"no-children-prop": no_children_prop_default,
		"no-children-prop-with-children": no_children_prop_with_children_default,
		"no-comment-textnodes": no_comment_textnodes_default,
		"no-key-after-spread": no_key_after_spread_default,
		"no-leaked-dollar": no_leaked_dollar_default,
		"no-leaked-semicolon": no_leaked_semicolon_default,
		"no-namespace": no_namespace_default,
		"no-useless-fragment": no_useless_fragment_default
	}
};

//#endregion
//#region src/configs/recommended.ts
var recommended_exports = /* @__PURE__ */ __exportAll({
	name: () => name$1,
	plugins: () => plugins$1,
	rules: () => rules$1,
	settings: () => settings$1
});
const name$1 = "react-jsx/recommended";
const rules$1 = {
	"react-jsx/no-children-prop": "warn",
	"react-jsx/no-children-prop-with-children": "error",
	"react-jsx/no-comment-textnodes": "warn",
	"react-jsx/no-key-after-spread": "error",
	"react-jsx/no-leaked-dollar": "warn",
	"react-jsx/no-leaked-semicolon": "warn",
	"react-jsx/no-namespace": "error"
};
const plugins$1 = { "react-jsx": plugin };
const settings$1 = { "react-x": DEFAULT_ESLINT_REACT_SETTINGS };

//#endregion
//#region src/configs/strict.ts
var strict_exports = /* @__PURE__ */ __exportAll({
	name: () => name,
	plugins: () => plugins,
	rules: () => rules,
	settings: () => settings
});
const name = "react-jsx/strict";
const rules = {
	...rules$1,
	"react-jsx/no-children-prop": "error",
	"react-jsx/no-useless-fragment": "warn"
};
const plugins = { ...plugins$1 };
const settings = { ...settings$1 };

//#endregion
//#region src/index.ts
const finalPlugin = {
	...plugin,
	configs: {
		["recommended"]: recommended_exports,
		["strict"]: strict_exports
	}
};

//#endregion
export { finalPlugin as default };
